<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Graphing of variables before classification &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="how-to, visualization">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Graphing of variables before classification">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Graphing of variables before classification">

<meta property="og:url" content="http://localhost:4000/2009/01/24/graphing-of-variables-before-classification/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2009/01/24/graphing-of-variables-before-classification/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#how-to" title="Pages tagged how-to">how-to</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#visualization" title="Pages tagged visualization">visualization</a></span>
        
          <h1 class="entry-title">Graphing of variables before classification</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2009-01-24T13:26:24-05:00"><i class="fa fa-calendar-o"></i> January 24, 2009</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <p>
It is important to be able to graphically evaluate variables that feed into classification algorithms. In addition to assessing the distribution of the data, visual inspection also reveals patterns that can be later tested statistically. This post describes the preparation of a quick histogram for data from positive and negative examples feeding into a classifier. The excellent <a href="http://matplotlib.sourceforge.net/">matplotlib</a> library is used for visualization.</p>
<p>
The example involves classifying two sets of proteins based on regional sequence charge. Two <a href="http://en.wikipedia.org/wiki/Fasta_format">FASTA</a> files were prepared, containing positive (active) and negative (non-active) examples. The goal is to look for a difference in charge between the two groups. Given a window size of amino acids to look at, we loop over the records in the file using the <a href="http://www.biopython.org/wiki/SeqIO">Biopython SeqIO</a> interface:</p>
<p>[sourcecode language="python"]<br />
def file_charges(in_file, cur_window):<br />
    all_charges = []<br />
    with open(in_file) as in_handle:<br />
        for rec in SeqIO.parse(in_handle, "fasta"):<br />
            cur_charges = calc_region_charges(rec.seq, cur_window)<br />
            all_charges.extend(cur_charges)<br />
    return all_charges<br />
[/sourcecode]</p>
<p>
We also use Biopython to calculate the <a href="http://en.wikipedia.org/wiki/Isoelectric_point">Isoelectric point</a> of each protein window. This is used as a proxy for charge; higher isoelectric points correspond to higher charges over the region.</p>
<p>[sourcecode language="python"]<br />
def calc_region_charges(seq, cur_window):<br />
    # internal small regions, so do not deal with C and N terminal charges<br />
    IsoelectricPoint.pKcterminal = {}<br />
    IsoelectricPoint.pKnterminal = {}<br />
    cur_pos = 0<br />
    region_charges = []<br />
    while cur_pos < len(seq) - cur_window:<br />
        cur_seq = seq[cur_pos:cur_pos + cur_window]<br />
        prot_analysis = ProtParam.ProteinAnalysis(str(cur_seq))<br />
        ie_calc = IsoelectricPoint.IsoelectricPoint(cur_seq,<br />
                prot_analysis.count_amino_acids())<br />
        region_charges.append(ie_calc.pi())<br />
        cur_pos += 1<br />
    return region_charges<br />
[/sourcecode]</p>
<p>
With this in place, we get the charges for the example records and plot them side by side as a histogram using matplotlib:</p>
<p>[sourcecode language="python"]<br />
cur_window = 75<br />
pos_charges = file_charges(pos_file, cur_window)<br />
neg_charges = file_charges(neg_file, cur_window)<br />
n, bins, patches = pylab.hist([pos_charges, neg_charges], 30,<br />
        normed=True, histtype='bar')<br />
pylab.xlabel('Isoelectric point')<br />
pylab.ylabel('Normalized percent of regions')<br />
pylab.title('Protein charge of %s amino acid windows' % cur_window)<br />
pylab.legend([patches[0][0], patches[1][0]], ['positives', 'negatives'])<br />
pylab.savefig('pos_neg_hist.png')<br />
pylab.show()<br />
[/sourcecode]</p>
<p>
The resulting graph shows the different distribution of charge between the positive and negative records. At an isoelectric point just above 10, only sequence windows from the positive examples are found. Looking at the percentage of highly charged sequence regions above this threshold in non-classified sequences could serve as the basis for a classifier.</p>
<p>[caption id="attachment_46" align="alignnone" width="400" caption="Example histogram"]<img src="http://bcbio.files.wordpress.com/2009/01/pos_neg_hist.png?w=400" alt="Example histogram" title="Histogram of example records" width="400" height="300" class="size-medium wp-image-46" />[/caption]</p>
<p>
The <a href="http://github.com/chapmanb/bcbb/blob/507cbb02c3531dddef67fd25087b3954f51e9d25/classify/histogram_thresh_find.py">example script</a> puts all of this together, and could be used as a template for classification problems in your own research.</p>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2009/01/19/finding-proteins-with-related-function-using-semantic-clustering/" class="btn" title="Finding proteins with related function using semantic clustering">Previous</a>
      
      
        <a href="http://localhost:4000/2009/01/31/location-and-duplication-information-from-ensembl/" class="btn" title="Location and duplication information from Ensembl">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
