<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>The influence of reduced resolution quality scores on alignment and variant calling &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="small-variants, germline, validation">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="The influence of reduced resolution quality scores on alignment and variant calling">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="The influence of reduced resolution quality scores on alignment and variant calling">

<meta property="og:url" content="http://localhost:4000/2013/02/13/the-influence-of-reduced-resolution-quality-scores-on-alignment-and-variant-calling/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2013/02/13/the-influence-of-reduced-resolution-quality-scores-on-alignment-and-variant-calling/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#small-variants" title="Pages tagged small-variants">small-variants</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#germline" title="Pages tagged germline">germline</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#validation" title="Pages tagged validation">validation</a></span>
        
          <h1 class="entry-title">The influence of reduced resolution quality scores on alignment and variant calling</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2013-02-13T05:49:00-05:00"><i class="fa fa-calendar-o"></i> February 13, 2013</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <div id="outline-container-1" class="outline-2">
<h2 id="sec-1">BAM file size reduction and quality score binning</h2>
<div class="outline-text-2" id="text-1">
<p> We have a large upcoming whole genome sequencing project with <a href="http://www.illumina.com/">Illumina</a>, and they approached us about delivering BAM files with reduced resolution base quality scores. They have a <a href="http://www.illumina.com/Documents/products/whitepapers/whitepaper_datacompression.pdf">white paper</a> describing the approach, which involves binning scores to reduce resolution.  This reduces the number of scores describing the quality of a base from 40 down to 8. </p>
<p> The advantage of this approach is a significant reduction in file size. BAM files use <a href="http://blastedbio.blogspot.ie/2011/11/bgzf-blocked-bigger-better-gzip.html">BGZF compression</a>, and the underlying <a href="http://www.chipestimate.com/tech-talks/2010/03/23/Altior-Unzipping-the-GZIP-compression-protocol">gzip DEFLATE algorithm</a> compresses based on shared text regions.  Reducing the number of quality values increases shared blocks and improves compression. This reduces BAM file sizes by 25-35%: an exome BAM file reduced from 5.7Gb to 3.7Gb after quality binning. </p>
<p> The potential downside is that the reduction in quality resolution may impact alignment and variant calling approaches that rely on base quality scores. To assess this, I implemented quality score binning as part of the <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen analysis pipeline</a> using the <a href="http://www.ebi.ac.uk/ena/about/cram_toolkit/">CRAM toolkit</a> and ran alignment, recalibration, realignment and variant calling on: </p>
<ul>
<li>The original unbinned 40-resolution base quality BAM from an   NA12878 exome. </li>
<li>The BAM binned into 8-resolution base qualities before   alignment. </li>
<li>The BAM binned into 8-resolution base qualities before alignment and   binned again following <a href="http://gatkforums.broadinstitute.org/discussion/44/base-quality-score-recalibration-bqsr">base quality score recalibration</a>. </li>
</ul>
<p> A comparison of alignment and variant calls from the three approaches indicates that binning has nearly no impact on alignment and a small impact on variant calls, primarily in low depth regions. </p>
</p></div>
</p></div>
<div id="outline-container-2" class="outline-2">
<h2 id="sec-2">Alignment differences</h2>
<div class="outline-text-2" id="text-2">
<p> We aligned 100bp paired end reads with <a href="http://www.novocraft.com/main/index.php">Novoalign</a>, a quality aware aligner. Comparison of mapped reads showed nearly no impact on total mapped reads. The plot below shows a generic delta of changes in mapped reads across the 22 autosomes alongside the increase in unmapped pairs. Out of 73 million total reads, the changes account for ~0.003% of the total reads. There also did not appear to be any worrisome patterns of loss for specific chromosomes. Overall, there is a minimal impact of quality score binning on the ability to align the reads. </p>
<p> <img src="http://bcbio.files.wordpress.com/2013/02/wpid-qualbin-alignment-changes.png" alt="Alignment changes following quality binning" width="600" /> </p>
</p></div>
</p></div>
<div id="outline-container-3" class="outline-2">
<h2 id="sec-3">Variant call differences</h2>
<div class="outline-text-2" id="text-3">
<p> We called variants using the <a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_genotyper_UnifiedGenotyper.html">GATK Unified Genotyper</a> following the <a href="http://gatkforums.broadinstitute.org/discussion/1186/best-practice-variant-detection-with-the-gatk-v4-for-release-2-0">best practice recommendations for exomes</a> and then compared calls from original and binned quality scores. Both approaches for binning -- pre-binning, and pre-binning plus post-quality recalibration binning -- showed similar levels of concordance to non-binned quality scores: 99.81 and 99.78, respectively. Since the additional binning after recalibration provides a smaller prepared BAM file for storage and has a similar impact to pre-binning only, we used this for additional analysis of discordant variants. </p>
<p> The table below shows the discordant differences between the 40 quality score resolution and binned, 8 quality score resolution BAMs. 40 quality discordant variants are those called with full quality score resolution but not called, or called differently, after binning to 8 quality score resolution. Conversely, the 8-quality discordants are those called uniquely after quality binning: </p>
<table border="1" cellspacing="0" cellpadding="6" rules="all" style="width:250px;margin:10px;">
<caption></caption>
<col class="left" />
<col class="right" />
<tbody>
<tr>
<td class="left">Overall genotype concordance</td>
<td class="right">99.78</td>
</tr>
<tr>
<td class="left">concordant: total</td>
<td class="right">117887</td>
</tr>
<tr>
<td class="left">concordant: SNPs</td>
<td class="right">109144</td>
</tr>
<tr>
<td class="left">concordant: indels</td>
<td class="right">8743</td>
</tr>
<tr>
<td class="left">40-quality discordant: total</td>
<td class="right">821</td>
</tr>
<tr>
<td class="left">40-quality discordant: SNPs</td>
<td class="right">759</td>
</tr>
<tr>
<td class="left">40-quality discordant: indels</td>
<td class="right">62</td>
</tr>
<tr>
<td class="left">8-quality discordant: total</td>
<td class="right">1289</td>
</tr>
<tr>
<td class="left">8-quality discordant: SNPs</td>
<td class="right">1240</td>
</tr>
<tr>
<td class="left">8-quality discordant: indels</td>
<td class="right">49</td>
</tr>
<tr>
<td class="left">het/hom discordant</td>
<td class="right">259</td>
</tr>
</tbody>
</table>
<p> We investigated the discordant variants further since 1.5%  of the total variant calls change as a result of binning, Of the 1851 unique discordant variants, approximately half (928) fall into reproducible variants  <a href="http://bcbio.wordpress.com/2013/02/06/an-automated-ensemble-method-for-combining-and-evaluating-genomic-variants-from-multiple-callers/">identified by looking at ensemble combinations of replicates</a>. Of these potentially problematic discordant variants more than half are in low coverage regions with less than 10 reads: </p>
<p> <img src="http://bcbio.files.wordpress.com/2013/02/wpid-qualbin-variant-changes.png" alt="Variant changes following quality binning" width="600" /> </p>
<p> The major influence of quality score binning is resolution of variants in low coverage regions. This manifests as differences in heterozygote and homozygote calling, indel representation and filtering differences related to quality and mappability. To assess the potential impact,  we looked at the loss in callable bases on a 30x whole genome sequence  when moving from a minimum of 5 reads to a minimum of 10, using  <a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_coverage_CallableLoci.html">GATK's CallableLoci tool</a>. Regions with read coverage of 5 to 9 make up 4.7 million genome positions, 0.17% of the  total callable bases. </p>
<table border="1" cellspacing="0" cellpadding="6" rules="all" style="width:400px;margin:10px;">
<caption></caption>
<col class="left" />
<col class="left" />
<col class="left" />
<thead>
<tr>
<th scope="col" class="left"></th>
<th scope="col" class="left">5 read minimum</th>
<th scope="col" class="left">10 read minimum</th>
</tr>
</thead>
<tbody>
<tr>
<td class="left">Callable bases</td>
<td class="left">2,775,871,235</td>
<td class="left">2,771,109,000</td>
</tr>
<tr>
<td class="left">Percent callable</td>
<td class="left">96.90%</td>
<td class="left">96.73%</td>
</tr>
<tr>
<td class="left">Low coverage</td>
<td class="left">17,641,980</td>
<td class="left">22,404,215</td>
</tr>
<tr>
<td class="left">No coverage/ poor mapping</td>
<td class="left">71,272,008</td>
<td class="left">71,272,008</td>
</tr>
</tbody>
</table>
<p> In conclusion, quality score binning provides a useful reduction in input file sizes with minimal impact on alignment. For variant calling, use additional caution in low coverage regions with less than 10 supporting reads. Given the rapid increases in read throughput that are driving the need for file size reduction, quality score binning is a worthwhile tradeoff for high-coverage recalling work. </p>
</p></div>
</p></div>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2013/02/06/an-automated-ensemble-method-for-combining-and-evaluating-genomic-variants-from-multiple-callers/" class="btn" title="An automated ensemble method for combining and evaluating genomic variants from multiple callers">Previous</a>
      
      
        <a href="http://localhost:4000/2013/04/08/bioinformatics-open-source-interoperability-hackathon-at-the-broad-institute/" class="btn" title="Bioinformatics open source interoperability Hackathon at the Broad Institute">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
