<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Validating generalized incremental joint variant calling with GATK HaplotypeCaller, FreeBayes, Platypus and samtools &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="joint-calling, germline, validation, small-variants">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Validating generalized incremental joint variant calling with GATK HaplotypeCaller, FreeBayes, Platypus and samtools">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Validating generalized incremental joint variant calling with GATK HaplotypeCaller, FreeBayes, Platypus and samtools">

<meta property="og:url" content="http://localhost:4000/2014/10/07/joint-calling/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2014/10/07/joint-calling/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#joint-calling" title="Pages tagged joint-calling">joint-calling</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#germline" title="Pages tagged germline">germline</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#validation" title="Pages tagged validation">validation</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#small-variants" title="Pages tagged small-variants">small-variants</a></span>
        
          <h1 class="entry-title">Validating generalized incremental joint variant calling with GATK HaplotypeCaller, FreeBayes, Platypus and samtools</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2014-10-07T08:53:00-04:00"><i class="fa fa-calendar-o"></i> October 07, 2014</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1">Incremental joint variant calling</h2>
<div class="outline-text-2" id="text-1">
<p> Variant calling in large populations is challenging due to the difficulty in providing a consistent set of calls at all possible variable positions. A finalized set of calls from a large population should distinguish reference calls, without a variant, from no calls, positions without enough read support to make a call. Calling algorithms should also be able to make use of information from other samples in the population to improve sensitivity and precision. </p>
<p> There are two issues with trying to provide complete combined call sets. First, it is computationally expensive to call a large number of samples simultaneously. Second, adding any new samples to a callset requires repeating this expensive computation. This <a href="http://gatkforums.broadinstitute.org/discussion/4150/should-i-analyze-my-samples-alone-or-together">N+1 problem</a> highlights the inflexibility around simultaneous pooled calling of populations. </p>
<p> The GATK team's recent 3.x release has a solution to these issues: <a href="https://www.broadinstitute.org/gatk/guide/article?id=3893">Incremental joint variant discovery</a>. The approach calls samples independently but produces a <a href="https://www.broadinstitute.org/gatk/guide/article?id=4017">genomic VCF (gVCF)</a> output for each individual that contains probability information for both variants and reference calls at non-variant positions. The genotyping step combines these individual gVCF files, making use of the information from the independent samples to produce a final callset. </p>
<p> We added GATK incremental joint calling to <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a> along with a <a href="https://github.com/chapmanb/bcbio.variation.recall">generalized implementation</a> that performs joint calling with other variant callers. Practically, bcbio now supports this approach with four variant callers: </p>
<ul class="org-ul">
<li><a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_gatk_tools_walkers_haplotypecaller_HaplotypeCaller.php">GATK HaplotypeCaller</a> (3.2-2) &#x2013; Follows current GATK recommended best practices for calling, with Variant Quality Score Recalibration used on whole genome and large population callsets. This uses individual sample gVCFs as inputs to joint calling.
</li>
<li><a href="https://github.com/ekg/freebayes">FreeBayes (0.9.14-15)</a> &#x2013; A haplotype-based caller from Erik Garrison in Gabor Marth's lab.
</li>
<li><a href="http://www.well.ox.ac.uk/platypus">Platypus (0.7.9.2)</a> &#x2013; A recently published haplotype-based variant caller from Andy Rimmer at the Wellcome Trust Centre for Human Genomics.
</li>
<li><a href="http://www.htslib.org/">samtools (1.0)</a> &#x2013; The recently released version of samtools and bcftools with a new <a href="http://samtools.github.io/bcftools/call-m.pdf">multiallelic calling method</a>. John Marshall, Petr Danecek, James Bonfield and Martin Pollard at Sanger have continued samtools development from Heng Li's code base.
</li>
</ul>
<p> The implementation includes integrated validation against the <a href="http://genomeinabottle.org/">Genome in a Bottle</a> NA12878 reference standard, allowing comparisons between joint calling, multi-sample pooled calling and single sample calling.  Sensitivity and precision for joint calling is comparable to pooled calling, suggesting we should optimize design of variant processing to cater towards individual calling and subsequent finalization of calls, rather than pooling. Generalized joint calling enables combining multiple sets of calls under an identical processing framework, which will be important as we seek to integrate large publicly available populations to extract biological signal in complex multi-gene diseases. </p>
</div>
</div>
<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2">Terminology</h2>
<div class="outline-text-2" id="text-2">
<p> There is not a consistent set of terminology around combined variant calling, but to develop one, here is how I'll use the terms: </p>
<ul class="org-ul">
<li>Joint calling &#x2013; Calling a group of samples together with algorithms that do not need simultaneous access to all population BAM files. GATK's incremental joint calling uses gVCF intermediates. Our generalized implementation performs recalling using individual BAMs supplemented with a combined VCF file of variants called in all samples.
</li>
<li>Pooled or batch calling &#x2013; Traditional grouped sample calling, where algorithms make use of read data from all BAM files of a group. This scales to smaller batches of samples.
</li>
<li>Single sample calling &#x2013; Variant calling with a single sample only, not making use of information from other samples.
</li>
<li>Squaring off or Backfilling &#x2013; Creating a VCF file from a group of samples that distinguishes reference from no-call at every position called as a variant in one of the samples. With a squared off VCF, we can use the <a href="http://cdn.vanillaforums.com/gatk.vanillaforums.com/FileUpload/9f/f0619642db06b73b599253f42ef2bf.png">sample matrix</a> to consider call rate at any position. Large populations called in smaller batches will not be able to distinguish reference from no-call at variants unique to each sub-pool, so will need to be re-processed to achieve this.
</li>
</ul>
</div>
</div>
<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3">Implementation</h2>
<div class="outline-text-2" id="text-3">
<p> <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a> automates the calling and validation used in this comparison. We aim to make it easy to install, use and extend. </p>
<p> For GATK HaplotypeCaller based joint genotyping, we implement the <a href="http://www.broadinstitute.org/gatk/guide/article?id=3893">GATK best practices</a> recommended by the Broad. Individual sample variant calls produce a gVCF output file that contains both variants as well as probability information about reference regions. Next, variants are jointly called using GenotypeGVFs to produce the final population VCF file. </p>
<p> For the other supported callers &#x2013; FreeBayes, Platypus and samtools &#x2013; we use a generalized recalling approach, implemented in <a href="https://github.com/chapmanb/bcbio.variation.recall">bcbio.variation.recall</a>. <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a> first calls each individual sample as a standard VCF. We then combine these individual sample VCFs into a global summary of all variant positions called across all samples. Finally we recall at each potential variant position, producing a globally squared off joint callset for the sample that we merge into the final joint VCF. This process parallelizes by chromosome region and by sample, allowing efficient use of resources in both clusters and <a href="http://jermdemo.blogspot.com/2011/06/big-ass-servers-and-myths-of-clusters.html">large multiple core machines</a>. </p>
<p> bcbio.variation.recall generalizes to any variant caller that supports recalling with an input set of variants. Knowing the context of potential variants helps inform better calling. This method requires having the individual sample BAM file available to perform recalling.  Having the reads present does provide the ability to improve recalling by taking advantage of realigning reads into haplotypes given known variants, an approach we'll explore more in future work. The implementation is also general and could support gVCF based combining as this becomes available for non-GATK callers. </p>
</div>
</div>
<div id="outline-container-sec-4" class="outline-2">
<h2 id="sec-4">Generalized joint calling</h2>
<div class="outline-text-2" id="text-4">
<p> We evaluated all callers against the NA12878 Genome in a Bottle reference standard using the <a href="http://ccr.coriell.org/Sections/Search/Sample_Detail.aspx?Ref=GM12878">NA12878/NA12891/NA12892 trio</a> from the <a href="http://blog.goldenhelix.com/wp-content/uploads/2013/03/Utah-Pedigree-1463-with-NA12878.png">CEPH 1463 Pedigree</a>, with 50x whole genome coverage from <a href="http://www.illumina.com/platinumgenomes/">Illumina's platinum genomes</a>. The validation provides putative true positives (concordant), false negatives (discordant missing), and false positives (discordant extra) for all callers: </p>
<p><a href="http://i.imgur.com/ddsRkkd.png"><br />
  <img src="http://i.imgur.com/ddsRkkd.png" width="650" alt="Incremental joint calling: GATK HaplotypeCaller, FreeBayes, Platypus, samtools" /><br />
</a></p>
<p> Overall, there is not a large difference in sensitivity and precision for the four methods, giving us four high-quality options for performing joint variant calling on germline samples. The post-calling filters provide similar levels of false positives to enable comparisons of sensitivity. Notably, samtools new calling method is now as good as other approaches, in contrast with <a href="http://bcbio.wordpress.com/2013/02/06/an-automated-ensemble-method-for-combining-and-evaluating-genomic-variants-from-multiple-callers/">previous evaluations</a>, demonstrating the value of continuing to improve open source tools and having updated benchmarks to reflect these improvements. </p>
<p> Improving sensitivity and precision is always an ongoing process and this evaluation identifies some areas to focus on for future work: </p>
<ul class="org-ul">
<li>Platypus SNP and indel calling is slightly less sensitive than other approaches. We worked on <a href="https://github.com/chapmanb/bcbio-nextgen/blob/9320479d8f21677b61ed1274b4da23d569c686ae/bcbio/variation/platypus.py#L29">Platypus calling parameters</a> and <a href="https://github.com/chapmanb/bcbio-nextgen/blob/9320479d8f21677b61ed1274b4da23d569c686ae/bcbio/variation/vfilter.py#L180">post-call filtering</a> to increase sensitivity from the defaults without introducing a large number of false positives, but welcome suggestions for more improvements.
</li>
<li>samtools indel calling needs additional work to reduce false positive indels in joint and pooled calling. There is more detail on this below in the comparison with single sample samtools calling.
</li>
</ul>
</div>
</div>
<div id="outline-container-sec-5" class="outline-2">
<h2 id="sec-5">Joint versus pooled versus single approaches</h2>
<div class="outline-text-2" id="text-5">
<p> We validated the same NA12878 trio with pooled and single sample calling to assess the advantages of joint calling over single sample, and whether joint calling is comparable in quality to calling simultaneously. The full evaluation for pooled calling shows that performance is similar to joint methods: </p>
<p><a href="http://i.imgur.com/Dna8hrI.png"><br />
  <img src="http://i.imgur.com/Dna8hrI.png" width="650" alt="Pooled calling: GATK HaplotypeCaller, FreeBayes, Platypus, samtools" /><br />
</a></p>
<p> If you plot joint, pooled and single sample calling next to each other there are some interesting small differences between approaches that identify areas for further improvement. As an example, here are GATK HaplotypeCaller and samtools with the three approaches presented side by side: </p>
<p><a href="http://i.imgur.com/bcCvXxP.png"><br />
  <img src="http://i.imgur.com/bcCvXxP.png" width="750" alt="Joint, pooled and single calling: GATK HaplotypeCaller and samtools" /><br />
</a></p>
<p> GATK HaplotypeCaller sensitivity and precision are close between the three methods, with small trade offs for different methods. For SNPs, pooled calling is most sensitive at the cost of more false positives, and single calling is more precise at the cost of some sensitivity. Joint calling is intermediate between these two extremes. For indels, joint calling is the most sensitive at the cost of more false positives, with pooled calling falling between joint and single sample calling. </p>
<p> For samtools, precision is currently best tuned for single sample calling. Pooled calling provides better sensitivity, but at the cost of a larger number of false positives. The joint calling implementation regains a bit of this sensitivity but still suffers from increased false positives. The authors of samtools tuned variant calling nicely for single samples, but there are opportunities to increase sensitivity when incorporating multiple samples via a joint method. </p>
<p> Generally, we don't expect the same advantages for pooled or joint calling in a trio as we'd see in a larger population. However, even for this small evaluation population we can see the improvements available by considering additional variant information from other samples. For Platypus we unexpectedly had better calls from joint calling compared to pooled calling, but expect these differences to harmonize over time as the tools continue to improve. </p>
<p> Overall, this comparison identifies areas where we can hope to improve generalized joint calling. We plan to provide specific suggestions and feedback to samtools, Platypus and other tool authors as part of a continuous validation and feedback process. </p>
</div>
</div>
<div id="outline-container-sec-6" class="outline-2">
<h2 id="sec-6">Reproducing and extending the analysis</h2>
<div class="outline-text-2" id="text-6">
<p> All variant callers and calling methods validated here are available for running in <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a>. bcbio automatically installs the generalized joint calling implementation, and it is also available as a java executable at <a href="https://github.com/chapmanb/bcbio.variation.recall">bcbio.variation.recall</a>. All tools are freely available, open source and community developed and we welcome your feedback and contributions. </p>
<p> The documentation contains <a href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/testing.html#whole-genome-trio-50x">full instructions for running the joint analysis</a>. This is an extended version of <a href="http://bcbio.wordpress.com/2014/05/12/wgs-trio-variant-evaluation/">previous work on validation of trio calling</a> and uses the same input dataset with a bcbio configuration that includes single, pooled and joint calling: </p>
<p>[sourcecode language="text" title=""]<br />
mkdir -p NA12878-trio-eval/config NA12878-trio-eval/input NA12878-trio-eval/work-joint<br />
cd NA12878-trio-eval/config<br />
cd ../input<br />
wget https://raw.github.com/chapmanb/bcbio-nextgen/master/config/examples/NA12878-trio-wgs-validate-getdata.sh<br />
bash NA12878-trio-wgs-validate-getdata.sh<br />
wget https://raw.github.com/chapmanb/bcbio-nextgen/master/config/examples/NA12878-trio-wgs-joint.yaml<br />
cd ../work_joint<br />
bcbio_nextgen.py ../config/NA12878-trio-wgs-joint.yaml -n 16<br />
[/sourcecode]<br /></p>
<p> Having a general joint calling implementation with good sensitivity and precision is a starting point for more research and development. To build off this work we plan to: </p>
<ul class="org-ul">
<li>Provide better <a href="http://bcbio.wordpress.com/2013/10/21/updated-comparison-of-variant-detection-methods-ensemble-freebayes-and-minimal-bam-preparation-pipelines/">ensemble calling methods</a> that scale to large multi-sample calling projects.
</li>
<li>Work with FreeBayes, Platypus and samtools tool authors to provide support for gVCF style files to avoid the need to have BAM files present during joint calling, and to improve sensitivity and precision during recalling-based joint approaches.
</li>
<li>Combine variant calls with local reassembly to improve sensitivity and precision. Erik Garrison's <a href="https://github.com/ekg/glia">glia</a> provides streaming local realignment given a set of potential variants. Jared Simpson used the <a href="https://github.com/jts/sga">SGA assembler</a> to combine <a href="https://github.com/jts/sga-extra">FreeBayes calls with de-novo assembly</a>. Ideally we could identify difficult regions of the genome <a href="http://knightlab.commons.yale.edu/gava-pt-2/">based on alignment information</a> and focus more computationally expensive assembly approaches there.
</li>
</ul>
<p> We plan to continue working with the open source scientific community to integrate, extend and improve these tools and are happy for any feedback and suggestions. </p>
</div>
</div>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2014/08/12/validated-whole-genome-structural-variation-detection-using-multiple-callers/" class="btn" title="Validated whole genome structural variation detection using multiple callers">Previous</a>
      
      
        <a href="http://localhost:4000/2014/12/19/awsbench/" class="btn" title="Benchmarking variation and RNA-seq analyses on Amazon Web Services with Docker">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
