<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Validating multiple cancer variant callers and prioritization in tumor-only samples &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="cancer, small-variants, structural-variation, validation">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Validating multiple cancer variant callers and prioritization in tumor-only samples">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Validating multiple cancer variant callers and prioritization in tumor-only samples">

<meta property="og:url" content="http://localhost:4000/2015/03/05/cancerval/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2015/03/05/cancerval/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#cancer" title="Pages tagged cancer">cancer</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#small-variants" title="Pages tagged small-variants">small-variants</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#structural-variation" title="Pages tagged structural-variation">structural-variation</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#validation" title="Pages tagged validation">validation</a></span>
        
          <h1 class="entry-title">Validating multiple cancer variant callers and prioritization in tumor-only samples</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2015-03-05T06:13:00-05:00"><i class="fa fa-calendar-o"></i> March 05, 2015</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1">Overview</h2>
<div class="outline-text-2" id="text-1">
<p> The post discusses work validating multiple cancer variant callers in <a href="http://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a> using a synthetic reference call set from <a href="https://www.synapse.org/#!Synapse:syn312572">the ICGC-TCGA DREAM challenge</a>.  We've previously <a href="http://bcb.io/2014/10/07/joint-calling/">validated germline variant calling methods</a>, but cancer calling is additionally challenging. Tumor samples have mixed cellularity due to contaminating normal sample, and consist of multiple sub-clones with different somatic variations. Low-frequency sub-clonal variations can be critical to understand disease progression but are more difficult to detect with high sensitivity and precision. </p>
<p> Publicly available whole genome truth sets like the <a href="http://genomeinabottle.org/">NA12878 Genome in a Bottle reference materials</a> don't currently exist for cancer calling, but other groups have been working to prepare standards for use in evaluating callers. The <a href="https://www.synapse.org/#!Synapse:syn312572">DREAM challenge</a> provides a set of <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/62018">synthetic datasets</a> that include cellularity and multiple sub-clones. There is also on ongoing DREAM contest with real, non-simulated data. In addition, the <a href="https://icgc.org/">ICGC</a> has done extensive work on <a href="http://biorxiv.org/content/early/2014/12/24/013177">assessing the challenges in cancer variant calling</a> and produced a detailed comparison of <a href="http://biorxiv.org/content/early/2014/12/24/013177">multiple variant calling pipelines</a>. Finally, <a href="http://www.bina.com/">Bina</a> recently released a <a href="https://github.com/bioinform/varsim">simulation framework called varsim</a> that they used to <a href="http://info.bina.com/hs-fs/hub/419826/file-1900487108-pdf/Posters/ASHG_2014_VarSim.pdf">compare multiple callers</a> as part of their <a href="http://info.bina.com/cancer-resources">cancer benchmarking</a>. I'm excited about all this community benchmarking and hope this work can contribute to the goal of having fully consented real patient reference materials, leading to a good understanding of variant caller tradeoffs. </p>
<p> In this work, we <a href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/testing.html#cancer-tumor-normal">evaluated cancer tumor/normal variant calling</a> with <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/62018">synthetic dataset 3</a> from the DREAM challenge, using multiple approaches to detect SNPs, indels and structural variants. A majority voting ensemble approach combines inputs from multiple callers into a final callset with good sensitivity and precision. We also provide a prioritization method to enrich for somatic mutations in tumor samples without matched normals. </p>
<p> Cancer variant calling in <a href="http://github.com/chapmanb/bcbio-nextgen">bcbio</a> is due to contributions and support from multiple members of the community: </p>
<ul class="org-ul">
<li><a href="https://github.com/mjafin">Miika Ahdesmaki</a> and <a href="https://twitter.com/BioInfo">Justin Johnson</a> at <a href="http://www.astrazeneca.com/Home">AstraZeneca</a> collaborated with our group on integrating and evaluating multiple variant callers. Their financial support helped fund our time on this comparison, and their scientific contributions improved SNP and indel calling.
</li>
<li><a href="https://github.com/lbeltrame">Luca Beltrame</a> integrated a number of cancer variant callers into bcbio and wrote the initial framework for somatic calling.
</li>
<li><a href="https://github.com/lpantano">Lorena Pantano</a> integrated variant callers and performed benchmarking work. Members of our group, the <a href="http://hsphbio.ghost.io/">Harvard Chan school Bioinformatics core</a>, regularly contribute to bcbio development and validation work.
</li>
<li>The <a href="http://www.gla.ac.uk/researchinstitutes/cancersciences/ics/">Wolfson Wohl Cancer Research Centre</a> supported our work on validation of somatic callers.
</li>
<li>James Cuff, Paul Edmon and the team at <a href="https://rc.fas.harvard.edu/">Harvard FAS research computing</a> provided compute infrastructure and support that enabled the large number of benchmarking runs it takes to get good quality calling.
</li>
</ul>
<p> I'm continually grateful to the community for all the contributions and support. The <a href="https://github.com/chapmanb/bcbio-nextgen/commits/master/bcbio/variation/mutect.py">MuTect commit history for bcbio</a> is a great example of multiple distributed collaborators working towards the shared goal of integrating and validating a caller. </p>
</div>
</div>
<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2">Variant caller validation</h2>
<div class="outline-text-2" id="text-2">
<p> We used a large collection of open source variant callers to detect SNPs, Indels and structural variants: </p>
<ul class="org-ul">
<li><a href="https://www.broadinstitute.org/cancer/cga/mutect">MuTect (1.1.5)</a> &#x2013; A SNP only caller, built on GATK UnifiedGenotyper, from the Broad Institute. MuTect requires a license if used for commercial purposes.
</li>
<li><a href="https://github.com/AstraZeneca-NGS/VarDict">VarDict (2014-12-15)</a> &#x2013; A SNP and indel caller from Zhongwu Lai and the oncology team at AstraZeneca.
</li>
<li><a href="https://github.com/ekg/freebayes">FreeBayes (0.9.20-1)</a> &#x2013; A haplotype aware realigning caller for SNPs and indels from Erik Garrison and Gabor Marth's lab.
</li>
<li><a href="http://varscan.sourceforge.net/">VarScan (2.3.7)</a> &#x2013; A heuristic/statistic based somatic SNP and indel caller from Dan Koboldt and The Genome Institute at Washington University.
</li>
<li><a href="http://scalpel.sourceforge.net/">Scalpel (0.3.1)</a> &#x2013; Micro-assembly based Indel caller from Giuseppe Narzisi and the Schatz lab. We pair Scalpel with MuTect to provide a complete set of small variant calls.
</li>
<li><a href="https://github.com/arq5x/lumpy-sv">LUMPY (0.2.7)</a> &#x2013; A <a href="http://genomebiology.com/2014/15/6/R84/abstract">probabilistic structural variant caller</a> incorporating both split read and read pair discordance, developed by Ryan Layer in <a href="http://quinlanlab.org/">Aaron Quinlan</a> and <a href="http://genome.wustl.edu/people/groups/detail/hall-lab/">Ira Hall's</a> labs.
</li>
<li><a href="https://github.com/tobiasrausch/delly">DELLY (0.6.1)</a> &#x2013; An <a href="http://bioinformatics.oxfordjournals.org/content/28/18/i333.abstract">integrated paired-end and split-end structural variant caller</a>   developed by Tobias Rausch.
</li>
<li><a href="https://github.com/jewmanchue/wham">WHAM (1.5.1)</a> &#x2013; A structural variant caller that can incorporate association testing, from Zev Kronenberg in Mark Yandell's lab at the University of Utah.
</li>
</ul>
<p> bcbio runs these callers and uses simple ensemble methods to combine <a href="http://bcb.io/2013/10/21/updated-comparison-of-variant-detection-methods-ensemble-freebayes-and-minimal-bam-preparation-pipelines/">small variants (SNPs, indels</a>) and <a href="http://bcb.io/2014/08/12/validated-whole-genome-structural-variation-detection-using-multiple-callers/">structural variants</a> into final combined callsets. The <a href="https://github.com/chapmanb/bcbio.variation.recall">new small variant ensemble method</a> uses a simplified majority rule classifier that picks variants to pass based on being present in a configurable number of samples. This performs well and is faster than the previous implementation that made use of both this approach and a subsequent support vector machine step. </p>
<p> Using the 100x whole genome tumor/normal pair from <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/62018">DREAM synthetic dataset 3</a> we evaluated each of the callers for sensitivity and precision on small variants (SNPs and indels). This synthetic dataset contains 100% tumor cellularity with 3 subclones at allele frequencies of 50%, 33% and 20%. </p>
<p><a href="http://i.imgur.com/Ck8AJQk.png"><br />
  <img src="http://i.imgur.com/Ck8AJQk.png" width="700" alt="Cancer evaluation for SNPs and indels" /><br />
</a></p>
<p> In addition to the whole genome results, the <a href="http://imgur.com/a/qba5k">validation album</a> includes results from running against the same dataset limited to exome regions. This has identical patterns of sensitivity and precision. It runs quicker, so is useful for evaluating changes to filtering or program parameters. </p>
<p> We also looked at structural variant calls for larger deletions, duplications and inversions. Here is the precision and sensitivity for duplications across multiple size classes: </p>
<p><a href="http://i.imgur.com/csQPl5p.png"><br />
  <img src="http://i.imgur.com/csQPl5p.png" width="700" alt="Cancer evaluation for structural variants -- duplications." /><br />
</a></p>
<p> The <a href="http://imgur.com/a/qba5k">full album of validation results</a> includes the comparisons for deletion and inversion events. These comparisons measure the contributions of individual callers within <a href="http://bcb.io/2014/08/12/validated-whole-genome-structural-variation-detection-using-multiple-callers/">an ensemble approach</a> that attempts to maximize sensitivity and specificity for the combined callset. Keep in mind that each of the individual results make use of other caller information in filtering. Our goal is to create the best possible combined calls, rather than a platform for unbiased comparison of structural variant callers. We're also actively working on improving sensitivity and specificity for individual callers and expect the numbers to continue to evolve. For example, <a href="https://github.com/jewmanchue/">Zev Kronenberg</a> added WHAM's ability to identify different classes of structural changes, and we're still in the process of improving the classifier. </p>
</div>
</div>
<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3">Improvements in filtering</h2>
<div class="outline-text-2" id="text-3">
<p> Our evaluation comparisons show best effort attempts to provide good quality calls for every caller. The final results often come from multiple rounds of improving sensitivity and precision by adjusting program parameters or downstream filtering. The goal of tightly integrating bcbio with validation is that the community can work on defining a set of parameters and filters that work best in multiple cases, and then use these directly within the same framework for processing real data. </p>
<p> In presenting the final results only, it may not be clear that plugging a specific tool into a custom bash script will not always produce the same results we see here. As an example, here are the improvements in FreeBayes sensitivity and precision from our initial implementation, presented over the exome regions of synthetic dataset 3: </p>
<p><a href="http://i.imgur.com/NJFaoas.png"><br />
  <img src="http://i.imgur.com/NJFaoas.png" width="500" alt="FreeBayes caller improvements" /><br />
</a></p>
<p> The original implementation used a <a href="https://github.com/ekg/vcflib">vcfsamplediff based</a> approach to filtering, as recommended on the <a href="https://groups.google.com/d/msg/freebayes/beLYRuHMkQE/RwFMniDmBYoJ">FreeBayes mailing list</a>. The current, improved, version uses a <a href="https://github.com/chapmanb/bcbio-nextgen/blob/4fe770cc1343f8e1a3f3fab1771bad13eb94df7a/bcbio/variation/freebayes.py#L217">custom filter based on genotype likelihoods</a>, based on the approach in the <a href="https://github.com/cc2qe/speedseq">speeseq pipeline</a>. </p>
<p> In general, you can find all of the integration work for individual callers in the <a href="https://github.com/chapmanb/bcbio-nextgen/tree/master/bcbio/variation">bcbio source code</a>, broken down by caller. For instance, here is the integration work on <a href="https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/mutect.py">MuTect</a>. The goal is to make all of the configuration settings and filters fully transparent so users can understand how they work when using bcbio, or transplant them into their own custom pipelines. </p>
</div>
</div>
<div id="outline-container-sec-4" class="outline-2">
<h2 id="sec-4">Tumor-only prioritization</h2>
<div class="outline-text-2" id="text-4">
<p> The above validations were all done on cancer calling with tumor and normal pairs. The filters to separate pre-existing <a href="https://en.wikipedia.org/wiki/Germline_mutation">germline mutations</a> from <a href="https://en.wikipedia.org/wiki/Mutation#Somatic_mutations">cancer specific somatic mutations</a> rely on the presence of a normal sample. In some cases, we don't have matched normal samples to do this filtering. Two common examples are <a href="http://www.mobio.com/blog/2012/07/20/formalin-fixed-paraffin-embedded-tissue-dna-isolation-the-basics/">FFPE samples</a> and tumor cell lines. For these samples, we'd like to be able to prioritize likely tumor specific variations for followup using publicly available resources. </p>
<p> We implemented a prioritization strategy from tumor-only samples in bcbio that takes advantage of publicly available resources like <a href="http://cancer.sanger.ac.uk/cancergenome/projects/cosmic/">COSMIC</a>, <a href="http://www.clinvar.com/">ClinVar</a>, <a href="http://www.1000genomes.org/">1000 genomes</a>, <a href="http://evs.gs.washington.edu/EVS/">ESP</a> and <a href="http://exac.broadinstitute.org/">ExAC</a>. It uses <a href="https://github.com/arq5x/gemini">GEMINI</a> to annotate the initial tumor-only VCF calls with external annotations, then extracts these to <a href="https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/prioritize.py">prioritize variants</a> with high or medium predicted impact, not present in 1000 genomes or ExAC at more than 1% in any subpopulation, or identified as pathenogenic in COSMIC or ClinVar. </p>
<p> Validating this prioritization strategy requires real tumor samples with known mutations. Our synthetic datasets are not useful here, since the variants <a href="https://www.synapse.org/#!Synapse:syn312572/wiki/62018">do not necessarily model standard biological variability</a>. You could spike in biologically relevant mutations, as done in the <a href="http://info.bina.com/hs-fs/hub/419826/file-1900487108-pdf/Posters/ASHG_2014_VarSim.pdf">VarSim cancer simulated data</a>, but this will bias towards our prioritization approach since both would use the same set of necessarily imperfect known variants and population level mutations. </p>
<p> We took the approach of using published tumor data with validated mutations. <a href="https://twitter.com/druvus">Andreas Sj√∂din</a> identified a <a href="http://onlinelibrary.wiley.com/doi/10.1002/hep.27243/full">Hepatoblastoma exome sequencing paper</a> with <a href="http://www.ebi.ac.uk/ena/data/view/SRP037747">publicly available sample data</a> and 23 validated cancer related variations across 5 samples. This is a baseline to help determine how stringent to be in removing potential germline variants. </p>
<p> The prioritization enriches variants of interest by 35-50x without losing sensitivity to confirmed variants: </p>
<table border="2" cellspacing="0" cellpadding="6" rules="groups">
<colgroup>
<col class="left" />
<col class="left" />
<col class="left" />
<col class="right" />
<col class="right" />
<col class="right" />
</colgroup>
<thead>
<tr>
<th scope="col" class="left">sample</th>
<th scope="col" class="left">caller</th>
<th scope="col" class="left">confirmed</th>
<th scope="col" class="right">enrichment</th>
<th scope="col" class="right">additional</th>
<th scope="col" class="right">filtered</th>
</tr>
</thead>
<tbody>
<tr>
<td class="left">HB2T</td>
<td class="left">freebayes</td>
<td class="left">6 / 7</td>
<td class="right">44x</td>
<td class="right">1288</td>
<td class="right">56046</td>
</tr>
<tr>
<td class="left">HB2T</td>
<td class="left">mutect</td>
<td class="left">6 / 7</td>
<td class="right">48x</td>
<td class="right">1014</td>
<td class="right">47755</td>
</tr>
<tr>
<td class="left">HB2T</td>
<td class="left">vardict</td>
<td class="left">6 / 7</td>
<td class="right">36x</td>
<td class="right">1464</td>
<td class="right">52090</td>
</tr>
<tr>
<td class="left">HB3T</td>
<td class="left">freebayes</td>
<td class="left">4 / 4</td>
<td class="right">46x</td>
<td class="right">1218</td>
<td class="right">54997</td>
</tr>
<tr>
<td class="left">HB3T</td>
<td class="left">mutect</td>
<td class="left">4 / 4</td>
<td class="right">49x</td>
<td class="right">961</td>
<td class="right">46894</td>
</tr>
<tr>
<td class="left">HB3T</td>
<td class="left">vardict</td>
<td class="left">4 / 4</td>
<td class="right">35x</td>
<td class="right">1511</td>
<td class="right">51404</td>
</tr>
<tr>
<td class="left">HB6T</td>
<td class="left">freebayes</td>
<td class="left">4 / 4</td>
<td class="right">43x</td>
<td class="right">1314</td>
<td class="right">56240</td>
</tr>
<tr>
<td class="left">HB6T</td>
<td class="left">mutect</td>
<td class="left">4 / 4</td>
<td class="right">51x</td>
<td class="right">946</td>
<td class="right">47747</td>
</tr>
<tr>
<td class="left">HB6T</td>
<td class="left">vardict</td>
<td class="left">3 / 4</td>
<td class="right">35x</td>
<td class="right">1497</td>
<td class="right">51625</td>
</tr>
<tr>
<td class="left">HB8T</td>
<td class="left">freebayes</td>
<td class="left">6 / 6</td>
<td class="right">42x</td>
<td class="right">1364</td>
<td class="right">57121</td>
</tr>
<tr>
<td class="left">HB8T</td>
<td class="left">mutect</td>
<td class="left">6 / 6</td>
<td class="right">47x</td>
<td class="right">1053</td>
<td class="right">48639</td>
</tr>
<tr>
<td class="left">HB8T</td>
<td class="left">vardict</td>
<td class="left">6 / 6</td>
<td class="right">35x</td>
<td class="right">1542</td>
<td class="right">52642</td>
</tr>
<tr>
<td class="left">HB9T</td>
<td class="left">freebayes</td>
<td class="left">2 / 2</td>
<td class="right">41x</td>
<td class="right">1420</td>
<td class="right">57582</td>
</tr>
<tr>
<td class="left">HB9T</td>
<td class="left">mutect</td>
<td class="left">2 / 2</td>
<td class="right">44x</td>
<td class="right">1142</td>
<td class="right">49858</td>
</tr>
<tr>
<td class="left">HB9T</td>
<td class="left">vardict</td>
<td class="left">2 / 2</td>
<td class="right">36x</td>
<td class="right">1488</td>
<td class="right">53098</td>
</tr>
</tbody>
</table>
<p> We consistently missed one confirmed mutation in the HB2T sample. This variant, reported as a somatic mutation in an <a href="http://useast.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000177673;r=2:231592901-231594283;t=ENST00000313965;redirect=no">uncharacterized open reading frame (C2orf57</a>), may actually be a germline mutation in the study sub-population. The variant is present at a <a href="http://exac.broadinstitute.org/variant/2-232458818-C-T">10% frequency in the East Asian population</a> but only 2% in the overall population, based on data from both the ExAC and 1000 genomes projects. Although the ethnicity of the original samples is not reported, the study authors are all from China. This helps demonstrate the effectiveness of large population frequencies, stratified by population, in prioritizing and evaluating variant calls. </p>
<p> The major challenge with tumor-only prioritization approaches is that you can't expect to accurately filter private germline mutations that you won't find in genome-wide catalogs. With a sample set using a small number of validated variants it's hard to estimate the proportion of 'additional' variants in the table above that are germline false positives versus the proportion that are additional tumor-only mutations not specifically evaluated in the study. We plan to continue to refine filtering with additional validated datasets to help improve this discrimination. </p>
<p> Practically, bcbio automatically runs prioritization with all tumor-only analyses. It filters variants by adding a ``LowPriority`` filter to the output VCF so users can readily identify variants flagged during the prioritization. </p>
</div>
</div>
<div id="outline-container-sec-5" class="outline-2">
<h2 id="sec-5">Future work</h2>
<div class="outline-text-2" id="text-5">
<p> This is a baseline for assessing SNP, indel and structural variant calls in cancer analyses. It also prioritizes impact variants in cases where we lack normal matched normals. We plan to continue to improve cancer variant calling in <a href="http://github.com/chapmanb/bcbio-nextgen">bcbio</a> and some areas of future focus include: </p>
<ul class="org-ul">
<li>Informing variant calling using estimates of tumor purity and sub-clonal frequency. bcbio integrates <a href="http://cnvkit.readthedocs.org/en/latest/">CNVkit</a>, a copy number caller, which exports read count segmentation data. Tools like <a href="https://github.com/raphael-group/THetA">THetA2</a>, <a href="https://github.com/morrislab/phylowgs">phyloWGS</a>, <a href="https://github.com/uci-cbcl/PyLOH">PyLOH</a> and <a href="https://github.com/genome/sciclone">sciClone</a> use these inputs to estimate normal contamination and sub-clonal frequencies.
</li>
<li>Focusing on difficult to call genomic regions and provide additional mechanisms to better resolve and improve caller accuracy in these regions. Using small variant calls to define problematic genome areas with large structural rearrangements can help prioritize and target the use of more computationally expensive methods for copy number assessment, structural variant calling and de-novo assembly.
</li>
<li>Evaluating calling and tumor-only prioritization on <a href="http://www.horizondx.com/our-products/q-seq-hdx-reference-standards">Horizon reference standards</a>. They contain a larger set of <a href="http://www.horizondx.com/media/item/202">validated mutations</a> at a variety of frequencies.
</li>
</ul>
<p> As always, we welcome suggestions, contributions and feedback. </p>
</div>
</div>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2014/12/19/awsbench/" class="btn" title="Benchmarking variation and RNA-seq analyses on Amazon Web Services with Docker">Previous</a>
      
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
