---
layout: post
title: Automated retrieval of expression data with Python and R
date: 2010-01-02 17:43:16.000000000 -05:00
categories:
- information retrieval
tags:
- how-to
status: publish
type: post
published: true
meta:
  _oembed_5375e464fc4edf83abb7bc29e046ba12: ! '{{unknown}}'
  _oembed_e2b57dbe08e556090504ec162343e160: ! '{{unknown}}'
  _oembed_96a738dcc5b38a5cc1ac02342e839c02: ! '{{unknown}}'
  _oembed_5bf41ef2fda75c9e7f7872fdd04fc08b: ! '{{unknown}}'
  _oembed_e76b2e985a0ea2a0700555fbdaf5ff0d: ! '{{unknown}}'
  _oembed_bec99ccc73ca7098147ce1e8aa54b56b: ! '{{unknown}}'
  _oembed_930c364a60be590b23c6f5119a563e8b: ! '{{unknown}}'
  _oembed_133a96193fb99659f0e9b1689a247d51: ! '{{unknown}}'
  _oembed_268df1a0bfe7c58ac254a304f93e6436: ! '{{unknown}}'
author: brad_chapman
excerpt: !ruby/object:Hpricot::Doc
---
<p>
In February, I'll be headed to Japan for <a href="http://hackathon3.dbcls.jp/">BioHackathon 2010</a>, which focuses on integrating biological datasets. In preparation, I've been thinking about use cases: when working with a set of data, what other information do you want to able to readily retrieve that will help you better answer your questions? One common set of data is gene expression information; you identify a group of genes of interest and are curious about their expression levels.</p>
<p>
In an ideal <a href="http://en.wikipedia.org/wiki/Semantic_Web">semantic</a> world, you would be able to build a query through a set of <a href="http://www.robertprice.co.uk/robblog/archive/2004/10/What_Is_An_RDF_Triple_.shtml">RDF triples</a> like:</p>
<ul>
<li> Expression set -&gt; has organism -&gt; Mus musculus
  </li>
<li> Mus musculus -&gt; has cell type -&gt; proB
  </li>
<li> proB -&gt; has biological state -&gt; wild type
  </li>
<li> Expression set -&gt; has identifier -&gt; A RefGene ID
  </li>
<li> A RefGene ID -&gt; has expression value -&gt; 7.65
</li>
</ul>
<p>
So you could find expression sets unique to your cell type under standard conditions, and then identify the expression values of your genes. Ideally, expression sets would also have additional meta-data associated with them so you'd know what 7.65 meant in the context of the experiment.</p>
<p>
Although we can't do this query through semantic channels, there are a number of toolkits that help us answer the question in an automated way. NCBI's Gene Expression Omnibus (<a href="http://www.ncbi.nlm.nih.gov/geo/">GEO</a>) provides an online resource hosting expression data. You can query the data sets with <a href="http://www.ncbi.nlm.nih.gov/entrez/utils/">EUtils</a> web API using <a href="http://biopython.org">Biopython</a>. Once a data set of interest is identified, Bioconductor's <a href="http://www.bioconductor.org/packages/bioc/html/GEOquery.html">GEOquery</a> can retrieve the data sets and mappings to UCSC RefGene identifiers. All of this is tied together with the <a href="http://rpy.sourceforge.net/rpy2.html">Rpy2</a> python interface to <a href="http://www.r-project.org/">R</a>.</p>
<p>
At the top level, we define our goal, which is to retrieve expression data for mouse proB wild type cells:</p>
<p>[sourcecode language="python"]<br />
def main():<br />
    organism = &quot;Mus musculus&quot;<br />
    cell_types = [&quot;proB&quot;, &quot;ProB&quot;, &quot;pro-B&quot;]<br />
    email = &quot;chapmanb@50mail.com&quot;<br />
    save_dir = os.getcwd()<br />
    exp_data = get_geo_data(organism, cell_types, email, save_dir,<br />
        _is_wild_type)</p>
<p>def _is_wild_type(result):<br />
    &quot;&quot;&quot;Check if a sample is wild type from the title.<br />
    &quot;&quot;&quot;<br />
    return result.samples[0][0].startswith(&quot;WT&quot;)<br />
[/sourcecode]</p>
<p>
We query GEO for the possible data sets based on our organism and cell type, and then search through the results for a dataset that matches our criteria. In real life work, your personal <code>_is_wild_type</code> function may be an interactive dialog that presents the biologist with possible experiments, allowing them to select the one of interest. Note also that we store our final result locally, as a Python <a href="http://docs.python.org/library/pickle.html">pickle</a>. Practically, this reduces our load on external servers by only querying them once:</p>
<p>[sourcecode language="python"]<br />
def get_geo_data(organism, cell_types, email, save_dir, is_desired_result):<br />
    save_file = os.path.join(save_dir, &quot;%s-results.pkl&quot; % cell_types[0])<br />
    if not os.path.exists(save_file):<br />
        results = cell_type_gsms(organism, cell_types, email)<br />
        for result in results:<br />
            if is_desired_result(result):<br />
                with open(save_file, &quot;w&quot;) as out_handle:<br />
                    cPickle.dump(result, out_handle)<br />
                break</p>
<p>    with open(save_file) as save_handle:<br />
        result = cPickle.load(save_handle)<br />
[/sourcecode]</p>
<p>
The hard work of querying GEO and retrieving the results is done by Biopython's Entrez interface. After building up the query, the results are parsed into simple objects with a description of the expression set along with titles and identifiers for each of the samples that match our cell type:</p>
<p>[sourcecode language="python"]<br />
def cell_type_gsms(organism, cell_types, email):<br />
    &quot;&quot;&quot;Use Entrez to retrieve GEO entries for an organism and cell type.<br />
    &quot;&quot;&quot;<br />
    Entrez.email = email<br />
    search_term = &quot;%s[ORGN] %s&quot; % (organism, &quot; OR &quot;.join(cell_types))<br />
    print &quot;Searching GEO and retrieving results: %s&quot; % search_term</p>
<p>    hits = []<br />
    handle = Entrez.esearch(db=&quot;gds&quot;, term=search_term)<br />
    results = Entrez.read(handle)<br />
    for geo_id in results['IdList']:<br />
        handle = Entrez.esummary(db=&quot;gds&quot;, id=geo_id)<br />
        summary = Entrez.read(handle)<br />
        samples = []<br />
        for sample in summary[0]['Samples']:<br />
            for cell_type in cell_types:<br />
                if sample['Title'].find(cell_type) &gt;= 0:<br />
                    samples.append((sample['Title'], sample['Accession']))<br />
                    break<br />
        if len(samples) &gt; 0:<br />
            hits.append(GEOResult(summary[0]['summary'], samples))<br />
    return hits<br />
[/sourcecode]</p>
<p>
Once the experiment is selected we've accomplished the first part of our goal. Next, the plan is to get expression data values. We do this by building up a dictionary that maps RefGene IDs to expression values, so the results look like:</p>
<p>[sourcecode language="python"]<br />
WT ProB, biological rep1<br />
[('NM_177327', [7.7430266269999999, 6.4795213670000003, 8.8766985500000004]),<br />
 ('NM_001008785', [7.4671954649999996, 5.4761453329999998]),<br />
 ('NM_177325', [7.3312364040000002, 11.831475960000001]),<br />
 ('NM_177323', [6.9779868059999997, 6.3926399939999996]),<br />
 ('NM_177322', [5.0833683379999997])]<br />
[/sourcecode]</p>
<p>
The actual hard work is retrieving the expression data and mappings to gene identifiers, and this is done using the GEOquery library in Bioconductor. Mapping information for expression values, and high level meta-data about the experiment, are stored in local files. The expression information is stored as a R style tab delimited data table, while the meta-data is stored in <a href="http://www.json.org/">JSON</a> format. Both sets of data are then present locally in readily readable formats for further processing:</p>
<p>[sourcecode language="python"]<br />
def _write_gsm_map(self, gsm_id, meta_file, table_file):<br />
    &quot;&quot;&quot;Retrieve GEO expression values using Bioconductor, saving to a table.<br />
    &quot;&quot;&quot;<br />
    robjects.r.assign(&quot;gsm.id&quot;, gsm_id)<br />
    robjects.r.assign(&quot;table.file&quot;, table_file)<br />
    robjects.r.assign(&quot;meta.file&quot;, meta_file)<br />
    robjects.r('''<br />
      library(GEOquery)<br />
      library(rjson)<br />
      gsm &lt;- getGEO(gsm.id)<br />
      write.table(Table(gsm), file = table.file, sep = &quot;\t&quot;, row.names = FALSE,<br />
                  col.names = TRUE)<br />
      cat(toJSON(Meta(gsm)), file = meta.file)<br />
    ''')<br />
[/sourcecode]</p>
<p>
The previous function provides a mapping between probe names and expression levels. To be useful, we also need to map each of the probes on the expression array to biological gene identifiers. This is done through a second call to the GEO library to retrieve details for the expression platform. Again the data is presented in a R data frame, which we limit to only the items of interest and save as a tab delimited file:</p>
<p>[sourcecode language="python"]<br />
def _write_gpl_map(self, gpl_id, gpl_file):<br />
    &quot;&quot;&quot;Retrieve GEO platform data using R and save to a table.<br />
    &quot;&quot;&quot;<br />
    robjects.r.assign(&quot;gpl.id&quot;, gpl_id)<br />
    robjects.r.assign(&quot;gpl.file&quot;, gpl_file)<br />
    robjects.r('''<br />
      library(GEOquery)<br />
      gpl &lt;- getGEO(gpl.id)<br />
      gpl.map &lt;- subset(Table(gpl), select=c(&quot;ID&quot;, &quot;RefSeq.Transcript.ID&quot;))<br />
      write.table(gpl.map, file = gpl.file, sep = &quot;\t&quot;, row.names = FALSE,<br />
                  col.names = TRUE)<br />
    ''')<br />
[/sourcecode]</p>
<p>
Putting this all together, we download each of the mapping files and then parse them, building the connection: expression to probe ID to gene transcript IDs. The final dictionary maps these gene identifiers to the expression values and returns them:</p>
<p>[sourcecode language="python"]<br />
def get_gsm_tx_values(self, gsm_id, save_dir):<br />
    &quot;&quot;&quot;Retrieve a map of transcripts to expression from a GEO GSM file.<br />
    &quot;&quot;&quot;<br />
    gsm_meta_file = os.path.join(save_dir, &quot;%s-meta.txt&quot; % gsm_id)<br />
    gsm_table_file = os.path.join(save_dir, &quot;%s-table.txt&quot; % gsm_id)<br />
    if (not os.path.exists(gsm_meta_file) or<br />
            not os.path.exists(gsm_table_file)):<br />
        self._write_gsm_map(gsm_id, gsm_meta_file, gsm_table_file)</p>
<p>    with open(gsm_meta_file) as in_handle:<br />
        gsm_meta = json.load(in_handle)<br />
    id_to_tx = self.get_gpl_map(gsm_meta['platform_id'], save_dir)<br />
    tx_to_vals = collections.defaultdict(list)<br />
    with open(gsm_table_file) as in_handle:<br />
        reader = csv.reader(in_handle, dialect='excel-tab')<br />
        reader.next() # header<br />
        for probe_id, probe_val in reader:<br />
            for tx_id in id_to_tx.get(probe_id, []):<br />
                tx_to_vals[tx_id].append(float(probe_val))<br />
    return tx_to_vals<br />
[/sourcecode]</p>
<p>
The <a href="http://github.com/chapmanb/bcbb/blob/master/rest_apis/find_geo_data.py">full script</a> puts all of these parts together into a working example. In the end we go from our query to the expression data we need, although the route is a bit more circuitous then the idealized one described at the start of the post. Simplifying this type of question based approach to data access is one challenge bioinformatics tool developers will continue to face.</p>
