<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Finding and displaying short reads clustered in the genome &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="how-to">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Finding and displaying short reads clustered in the genome">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Finding and displaying short reads clustered in the genome">

<meta property="og:url" content="http://localhost:4000/2009/04/29/finding-and-displaying-short-reads-clustered-in-the-genome/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2009/04/29/finding-and-displaying-short-reads-clustered-in-the-genome/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#how-to" title="Pages tagged how-to">how-to</a></span>
        
          <h1 class="entry-title">Finding and displaying short reads clustered in the genome</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2009-04-29T16:55:06-04:00"><i class="fa fa-calendar-o"></i> April 29, 2009</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <p>
<a href="http://en.wikipedia.org/wiki/Next-generation_sequencing#New_sequencing_methods">Short read, or next-generation, sequencing</a> is becoming increasingly prevalent in the day-to-day lives of bioinformatics researchers. Programs such as <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a>, <a href="http://apps.sourceforge.net/mediawiki/cloudburst-bio/index.php?title=CloudBurst">CloudBurst</a>, <a href="http://maq.sourceforge.net/">Maq</a>, and <a href="http://www.broad.mit.edu/crd/wiki/index.php/Arachne">Arachne</a> are available to align short reads to a reference genome. Once you have these alignments, the data analysis work begins. During a recent project, I was interested in looking at overlapping read groups across a genome; this post describes a method for collecting and visualizing those groups.</p>
<p>
Starting with a file from an alignment program, the first step is to write a <a href="http://www.python.org/dev/peps/pep-0255/">Python generator</a> that returns information about the alignments. Since there are many different aligners, below is some Python pseudocode which describes the function:</p>
<p>[sourcecode language="python"]<br />
def parse_alignment(align_file, errors_allowed=0):<br />
    with open(align_file) as align_handle:<br />
        for (read_id, match_id, match_start, match_end, errors) in \<br />
                your_parser.read_iterator(align_handle):<br />
            if (errors &lt;= errors_allowed):<br />
                yield read_id, match_id, match_start, match_end<br />
[/sourcecode]</p>
<p>
It parses a line, checks if it has an acceptable number of alignment errors, and yields a unique id for the read, an id for the match like a chromosome or contig name, and the start and end coordinates of the match. We can use this generator to build our representation of overlapping reads with the help of the excellent <a href="http://bitbucket.org/james_taylor/bx-python/wiki/Home">bx-python</a> library. bx-python contains a Python and C implementation of clustered <a href="http://hackmap.blogspot.com/2008/11/python-interval-tree.html">interval trees</a>. The function below uses the interface to generate <code>ClusterTree</code> objects for each chromosome/contig in your alignment:</p>
<p>[sourcecode language="python"]<br />
import collections<br />
from bx.intervals.cluster import ClusterTree</p>
<p>def build_cluster_trees(align_generator, cluster_distance):<br />
    # arguments to ClusterTree are:<br />
    # - Distance in basepairs for two reads to be in the same cluster;<br />
    #   for instance 20 would group all reads with 20bp of each other<br />
    # - Number of reads necessary for a group to be considered a cluster;<br />
    #   2 returns all groups with 2 or more overlapping reads<br />
    cluster_trees = collections.defaultdict(lambda:<br />
            ClusterTree(cluster_distance, 2))<br />
    for read_id, match_id, start, end in align_generator:<br />
        cluster_trees[match_id].insert(start, end, read_id)<br />
    return dict(cluster_trees)<br />
[/sourcecode]</p>
<p>
The generated trees will readily provide a list of start and end coordinates for a clustered region, along with all of the reads that map to that region:</p>
<p>[sourcecode language="python"]<br />
for chrom, cluster_tree in cluster_trees.items():<br />
    for start, end, read_ids in cluster_tree.getregions():<br />
        # do something with your read cluster<br />
[/sourcecode]</p>
<p>
My interest was in visualizing the reads in these clusters along with the frequency of each read. To do so, I generated two python dictionaries which map the <code>read_id</code> identifiers, used in the grouping, to the chromosome location of the read (<code>location_db</code>) and the number of times a read was found in the raw short read results (<code>freq_db</code>). Practically, these are implemented as <a href="http://www.jcea.es/programacion/pybsddb.htm">BerkeleyDB</a> key value stores, to avoid having to keep all of the read information in memory.</p>
<p>
With this information, we can use <a href="http://matplotlib.sourceforge.net/">matplotlib</a> to visualize the reads. Frequencies are represented using a <a href="http://www.astro.princeton.edu/~msshin/science/code/matplotlib_cm/">color map</a>. Each read in a group is plotted using horizontal bar graphs:</p>
<p>[sourcecode language="python"]<br />
import pylab</p>
<p>def plot_reads(cluster_name, read_ids, location_db, freq_db):<br />
    read_info = []<br />
    for read_id in read_ids:<br />
        start, end = location_db[read_id]<br />
        freq = freq_db[read_id]<br />
        read_info.append((start, end, freq))<br />
    read_info.sort(reverse=True)<br />
    min_freq = min(l[2] for l in read_info)<br />
    max_freq = max(l[2] for l in read_info)<br />
    freq_cmap = pylab.get_cmap('Blues')<br />
    pylab.clf()<br />
    for rindex, (start, end, freq) in enumerate(read_info):<br />
        freq_percent = float(freq - min_freq) / float(max_freq - min_freq)<br />
        color = freq_cmap(freq_percent)<br />
        if freq in [min_freq, max_freq]:<br />
            label = "Frequency: %s" % (freq)<br />
        else:<br />
            label = None<br />
        if label and label not in labels_done:<br />
            labels_done.append(label)<br />
        else:<br />
            label = None<br />
        pylab.barh(rindex, end - start, left=start, height=0.8, color=color,<br />
                label=label)<br />
    pylab.title(cluster_name)<br />
    pylab.xlabel("Coordinates")<br />
    pylab.ylabel("Reads")<br />
    pylab.legend(loc='upper right')<br />
    pylab.yticks(())<br />
    out_file = "%s.png" % (cluster_name)<br />
    pylab.savefig(out_file)<br />
[/sourcecode]</p>
<p>
The resulting graph for an example region provides a quick visual summary of read coverage, overlap, and reliability:</p>
<p>[caption id="attachment_95" align="alignnone" width="400" caption=""]<img src="http://bcbio.files.wordpress.com/2009/04/5_5429173_5429347-freq.png?w=300" alt="Example coverage plot" title="5_5429173_5429347-freq" width="400" height="300" class="size-medium wp-image-65" />[/caption]<br /></p>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2009/04/20/biopython-projects-for-google-summer-of-code/" class="btn" title="Biopython projects for Google Summer of Code">Previous</a>
      
      
        <a href="http://localhost:4000/2009/05/10/evaluating-key-value-and-document-stores-for-short-read-data/" class="btn" title="Evaluating key-value and document stores for short read data">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
