<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Organization of literature using PubMed related articles &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="how-to">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Organization of literature using PubMed related articles">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Organization of literature using PubMed related articles">

<meta property="og:url" content="http://localhost:4000/2009/02/13/organization-of-literature-using-pubmed-related-articles/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2009/02/13/organization-of-literature-using-pubmed-related-articles/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#how-to" title="Pages tagged how-to">how-to</a></span>
        
          <h1 class="entry-title">Organization of literature using PubMed related articles</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2009-02-13T06:19:05-05:00"><i class="fa fa-calendar-o"></i> February 13, 2009</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <p>
When dealing with a long list of journal articles, what is the best method to organize them? I was confronted with this problem in designing an interface where users would pick specific papers and retrieve results tied to them. Presenting them as the raw list was unsatisfying; it is fine for users who know exactly what articles they want, but naive users would have a lot of difficulty finding relevant articles. Even for power users, a better classification system could help reveal items they may not have known about.</p>
<p>
The approach I took was to group together papers based on similarity. The <a href="http://www.ncbi.nlm.nih.gov/sites/entrez?db=PubMed">NCBI PubMed</a> literature database has links to related articles, which it exposes programmatically though <a href="http://www.ncbi.nlm.nih.gov/entrez/query/static/eutils_help.html">EUtils</a>. Using the <a href="http://biopython.org/DIST/docs/tutorial/Tutorial.html">Biopython</a> Entrez interface, the first step is to retrieve a dictionary of related IDs for each starting article, ordered by relevance:</p>
<p>[sourcecode language="python"]<br />
def _get_elink_related_ids(self, pmids):<br />
    pmid_related = {}<br />
    for pmid in pmids:<br />
        handle = Entrez.elink(dbform='pubmed', db='pubmed', id=pmid)<br />
        record = Entrez.read(handle)<br />
        cur_ids = []<br />
        for link_dict in record[0]['LinkSetDb'][0]['Link']:<br />
            cur_ids.append((int(link_dict.get('Score', 0)),<br />
                link_dict['Id']))<br />
        cur_ids.sort()<br />
        cur_ids.reverse()<br />
        local_ids = [x[1] for x in cur_ids if x[1] in pmids]<br />
        if pmid in local_ids:<br />
            local_ids.remove(pmid)<br />
        pmid_related[pmid] = local_ids<br />
    return pmid_related<br />
[/sourcecode]</p>
<p>
Trying to group directly based on this dictionary will often result in one large group, since many of the articles may be linked together through a few common articles. For instance, a review may be related to several other papers in non-overlapping areas. To make the results as useful as possible we define a maximum and minimum group size, and a two parameters to filter the related lists:</p>
<ul>
<li><code>overrep_thresh</code>: The percentage of papers an item is related to out of all papers being grouped; the threshold sets a maximum number of papers that can be related. For instance, a value of .25 means that a journal will be related to 25% or less of the total papers.
  </li>
<li><code>related_max</code>: The number of related papers to use. The best related articles go into the grouping.
</li>
</ul>
<p>
These parameters define a filter for our dictionary of related articles:</p>
<p>[sourcecode language="python"]<br />
def _filter_related(self, inital_dict, overrep_thresh, related_max):<br />
    final_dict = {}<br />
    all_vals = reduce(operator.add, inital_dict.values())<br />
    for item_id, item_vals in inital_dict.items():<br />
        final_vals = [val for val in item_vals if<br />
            float(all_vals.count(val)) / len(inital_dict) <= overrep_thresh]<br />
        final_dict[item_id] = final_vals[:related_max]<br />
    return final_dict<br />
[/sourcecode]</p>
<p>
The filtered list is grouped using a generalized version of the <code>examine_paralogs</code> function used in an earlier post to group together <a href="http://bcbio.wordpress.com/2009/01/31/location-and-duplication-information-from-ensembl/">location and duplication information</a>. Sets combine any groups with overlapping articles:</p>
<p>[sourcecode language="python"]<br />
def _groups_from_related_dict(self, related_dict):<br />
    cur_groups = []<br />
    all_base = related_dict.keys()<br />
    for base_id, cur_ids in related_dict.items():<br />
        overlap = set(cur_ids) & set(all_base)<br />
        if len(overlap) > 0:<br />
            new_group = set(overlap | set([base_id]))<br />
            is_unique = True<br />
            for exist_i, exist_group in enumerate(cur_groups):<br />
                if len(new_group & exist_group) > 0:<br />
                    update_group = new_group | exist_group<br />
                    cur_groups[exist_i] = update_group<br />
                    is_unique = False<br />
                    break<br />
            if is_unique:<br />
                cur_groups.append(new_group)<br />
    return [list(g) for g in cur_groups]<br />
[/sourcecode]</p>
<p>
With this list, we want to extract the groups and their articles that fit in our grouping criteria, the minimum and maximum size:</p>
<p>[sourcecode language="python"]<br />
def _collect_new_groups(self, pmid_related, groups):<br />
    final_groups = []<br />
    for group_items in groups:<br />
        final_items = [i for i in group_items if pmid_related.has_key(i)]<br />
        if (len(final_items) >= self._min_group and<br />
                len(final_items) <= self._max_group):<br />
            final_groups.append(final_items)<br />
            for item in final_items:<br />
                del pmid_related[item]<br />
    final_related_dict = {}<br />
    for pmid, related in pmid_related.items():<br />
        final_related = [r for r in related if pmid_related.has_key(r)]<br />
        final_related_dict[pmid] = final_related<br />
    return final_groups, final_related_dict<br />
[/sourcecode]</p>
<p>
Utilizing these functions, the main algorithm steps through a series of increasingly less stringent parameters picking out groups which fall into our thresholds. Closely related journal articles are grouped first; more general papers with less association will be placed in groups in later rounds:</p>
<p>[sourcecode language="python"]<br />
def get_pmid_groups(self, pmids):<br />
    pmid_related = self._get_elink_related_ids(pmids)<br />
    filter_params = self._filter_params[:]<br />
    final_groups = []<br />
    while len(pmid_related) > 0:<br />
        if len(filter_params) == 0:<br />
            raise ValueError("Ran out of parameters before finding groups")<br />
        cur_thresh, cur_related = filter_params.pop(0)<br />
        while 1:<br />
            filt_related = self._filter_related(pmid_related, cur_thresh,<br />
                    cur_related)<br />
            groups = self._groups_from_related_dict(filt_related)<br />
            new_groups, pmid_related = self._collect_new_groups(<br />
                    pmid_related, groups)<br />
            final_groups.extend(new_groups)<br />
            if len(new_groups) == 0:<br />
                break<br />
        if len(pmid_related) < self._max_group:<br />
            final_groups.append(pmid_related.keys())<br />
            pmid_related = {}<br />
    return final_groups<br />
[/sourcecode]</p>
<p>
The full code wrapped up into a class is available from <a href="http://github.com/chapmanb/bcbb/blob/da44ff1329f66188edd4da3aca4b93c18c216195/classify/pubmed_related_group.py">the GitHub repository</a>.</p>
<p>
This is one approach to automatically grouping a large list of literature to make interesting items more discoverable. With the work being done on full text indexing, the data underlying resources such as <a href="http://www.ihop-net.org/UniPub/iHOP/">iHOP</a> might be used to do these groupings even more effectively using similar algorithms.</p>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2009/02/07/automated-protein-conservation-display-from-blast-alignments/" class="btn" title="Automated protein conservation display from BLAST alignments">Previous</a>
      
      
        <a href="http://localhost:4000/2009/02/22/exploring-bioperl-genbank-to-gff-mapping/" class="btn" title="Exploring BioPerl GenBank to GFF mapping">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
