<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Whole genome trio variant calling evaluation: low complexity regions, GATK VQSR and high depth filters &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="small-variants, validation">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Whole genome trio variant calling evaluation: low complexity regions, GATK VQSR and high depth filters">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Whole genome trio variant calling evaluation: low complexity regions, GATK VQSR and high depth filters">

<meta property="og:url" content="http://localhost:4000/2014/05/12/wgs-trio-variant-evaluation/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2014/05/12/wgs-trio-variant-evaluation/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#small-variants" title="Pages tagged small-variants">small-variants</a>&nbsp;&bull;&nbsp;<a href="http://localhost:4000/tags/#validation" title="Pages tagged validation">validation</a></span>
        
          <h1 class="entry-title">Whole genome trio variant calling evaluation: low complexity regions, GATK VQSR and high depth filters</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2014-05-12T06:03:22-04:00"><i class="fa fa-calendar-o"></i> May 12, 2014</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <div id="outline-container-sec-1" class="outline-2">
<h2 id="sec-1">Whole genome trio validation</h2>
<div class="outline-text-2" id="text-1">
<p> I've written previously about the approaches we use to validate the <a href="https://github.com/chapmanb/bcbio-nextgen">bcbio-nextgen</a> variant calling framework, specifically evaluating <a href="https://bcbio.wordpress.com/2013/05/06/framework-for-evaluating-variant-detection-methods-comparison-of-aligners-and-callers/">aligners and variant calling methods</a> and <a href="https://bcbio.wordpress.com/2013/10/21/updated-comparison-of-variant-detection-methods-ensemble-freebayes-and-minimal-bam-preparation-pipelines/">assessing the impact of BAM post-alignment preparation methods</a>. We're continually looking to improve both the pipeline and validation methods and two recent papers helped advance best-practices for evaluating and filtering variant calls: </p>
<ul class="org-ul">
<li><a href="http://research.mssm.edu/linderman/index.html">Michael Linderman and colleagues</a> describe approaches for <a href="http://www.biomedcentral.com/1755-8794/7/20/abstract">validating clinical exome and whole genome sequencing results</a>. One key result I took from the paper was the difference in assessment between exome and whole genome callsets. Coverage differences due to capture characterize discordant exome variants, while complex genome regions drive whole genome discordants. Reading this paper pushed us to evaluate whole genome population based variant calling, which is now feasible due to improvements in bcbio-nextgen scalability. </li>
<li><a href="https://twitter.com/lh3lh3">Heng Li</a> identified <a href="http://arxiv.org/abs/1404.0929">variant calling artifacts and proposed filtering approaches</a> to remove them, as well as characterizing caller error rates. We'll investigate two of the filters he proposed: removing variants in low complexity regions, and filtering high depth variants. </li>
</ul>
<p> We use the <a href="http://ccr.coriell.org/Sections/Search/Sample_Detail.aspx?Ref=GM12878">NA12878/NA12891/NA12892 trio</a> from the <a href="http://blog.goldenhelix.com/wp-content/uploads/2013/03/Utah-Pedigree-1463-with-NA12878.png">CEPH 1463 Pedigree</a> as an input dataset, consisting of 50x whole genome reads from <a href="http://www.illumina.com/platinumgenomes/">Illumina's platinum genomes</a>. This enables both whole genome comparisons, as well as pooled family calling that replicates best-practice for calling within populations. We aligned reads using <a href="http://bio-bwa.sourceforge.net/">bwa-mem</a> and performed streaming de-duplication detection with <a href="https://github.com/GregoryFaust/samblaster">samblaster</a>. Combined with no recalibration or realignment based on <a href="https://bcbio.wordpress.com/2013/10/21/updated-comparison-of-variant-detection-methods-ensemble-freebayes-and-minimal-bam-preparation-pipelines/">our previous assessment</a>, this enabled fully streamed preparation of BAM files from input fastq reads. We called variants using two realigning callers: <a href="https://github.com/ekg/freebayes">FreeBayes (v0.9.14-7)</a> and <a href="http://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_haplotypecaller_HaplotypeCaller.html">GATK HaplotypeCaller (3.1-1-g07a4bf8)</a> and evaluated calls using the <a href="http://www.nature.com/nbt/journal/v32/n3/full/nbt.2835.html">Genome in a Bottle reference callset for NA12878 (v0.2-NIST-RTG)</a>. The bcbio-nextgen documentation has <a href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/testing.html#whole-genome-trio-50x">full instructions for reproducing the analysis</a>. </p>
<p> This work provides three practical improvements for variant calling and validation: </p>
<ul class="org-ul">
<li>Low complexity regions contribute 45% of the indels in whole genome evaluations, and are highly variable between callers. This replicates Heng's results and Michael's assessment of common errors in whole genome samples, and indicates we need to specifically identify and assess the 2% of the genome labeled as low complexity. Practically, we'll exclude them from further evaluations to avoid non-representative bias, and suggest removing or flagging them when producing whole genome variant calls. </li>
<li>We add a filter for removing false positives from FreeBayes calls in high depth, low quality regions. This removes variants in high depth regions that are likely due to copy number or other larger structural events, and replicates Heng's filtering results. </li>
<li>We improved settings for <a href="http://gatkforums.broadinstitute.org/discussion/39/variant-quality-score-recalibration-vqsr">GATK variant quality recalibration (VQSR)</a>. The default VQSR settings are conservative for SNPs and need adjustment to be compatible with the sensitivity available through FreeBayes or GATK using hard filters. </li>
</ul></div>
</p></div>
<div id="outline-container-sec-2" class="outline-2">
<h2 id="sec-2">Low complexity regions</h2>
<div class="outline-text-2" id="text-2">
<p> Low complexity regions (LCRs) consist of locally repetitive sections of the genome. Heng's paper identified these using <a href="http://compbio.dfci.harvard.edu/tgi/software/">mdust</a> and provides a <a href="https://github.com/lh3/varcmp/raw/master/scripts/LCR-hs37d5.bed.gz">BED file of LCRs</a> covering 2% of the genome. Repeats in these regions can lead to artifacts in sequencing and variant calling. Heng's paper provides examples of areas where a full de-novo assembly correctly resolves the underlying structure, while local reassembly variant callers do not. </p>
<p> To assess the impact of low complexity regions on variant calling, we compared calls from FreeBayes and GATK HaplotypeCaller to the Genome in a Bottle reference callset with and without low complexity regions included. The graph below shows concordant non-reference variant calls alongside discordant calls in three categories: missing discordants are potential false negatives, extras are potential false positives, and shared are variants that overlap between the evaluation and reference callsets but differ due to zygosity (heterozygote versus homozygote) or indel representation. </p>
<p>  <a href="http://i.imgur.com/D2z9Kyp.png">   <img src="http://i.imgur.com/D2z9Kyp.png" width="700" alt="Low complexity regions for GATK and FreeBayes validation" /> </a>
<ul class="org-ul">
<li>For SNPs, removing low complexity regions removes approximately ~2% of the total calls for both FreeBayes and GATK. This corresponds to the 2% of the genome subtracted by removing LCRs. </li>
<li>For indels, removing LCRs removes 45% of the calls due to the over-representation of indels in repeat regions. Additionally, this results in approximately equal GATK and FreeBayes concordant indels after LCR removal. Since the Genome in a Bottle reference callset uses GATK HaplotypeCaller to resolve discrepant calls, this change in concordance is likely due to bias towards GATK's approaches for indel resolution in complex regions. </li>
<li>The default GATK VQSR calls for SNPs are not as sensitive, relative to FreeBayes calls. I'll describe additional work to improve this below. </li>
</ul>
<p> Practically, we'll now exclude low complexity regions in variant comparisons to avoid potential bias and more accurately represent calls in the remaining non-LCR genome. We'll additionally flag low complexity indels in non-evaluation callsets as likely to require additional followup. Longer term, we need to incorporate callers specifically designed for repeats like <a href="http://lobstr.teamerlich.org/index.html">lobSTR</a> to more accurately characterize these regions. </p>
</p></div>
</p></div>
<div id="outline-container-sec-3" class="outline-2">
<h2 id="sec-3">High depth, low quality, filter for FreeBayes</h2>
<div class="outline-text-2" id="text-3">
<p> The second filter proposed in Heng's paper was removal of high depth variants. This was a useful change in mindset for me as I've primarily thought about removing low quality, low depth variants. However, high depth regions can indicate potential copy number variations or hidden duplicates which result in spurious calls. </p>
<p> Comparing true and false positive FreeBayes calls with a pooled multi-sample call quality of less than 500 identifies a large grouping of false positive heterozygous variants at a combined depth, across the trio, of 200: </p>
<p>  <a href="http://i.imgur.com/S9lObRf.png">   <img src="http://i.imgur.com/S9lObRf.png" width="700" alt="Heterozygotes by depth and quality: true versus false positive" /> </a>
<p> The cutoff proposed by Heng was to calculate the average depth of called variants and set the cutoff as the average depth plus a multiplier of 3 or 4 times the square root of average depth. This dataset was an average depth of 169 for the trio, corresponding to a cutoff of 208 if we use the 3 multiplier, which compares nicely with a manual cutoff you'd set looking at the above graphs. Applying a cutoff of QUAL &lt; 500 and DP &gt; 208 produces a reduction in false positives with little impact on sensitivity: </p>
<p>  <a href="http://i.imgur.com/zOcrnKS.png">   <img src="http://i.imgur.com/zOcrnKS.png" width="700" alt="Improvement in filtering false positives with high depth filter" /> </a>
<p> A nice bonus of this filter is that it makes intuitive sense: variants with high depth and low quality indicate there is something problematic, and depth manages to partially compensate for the underlying issue. Inspired by <a href="https://www.broadinstitute.org/gatk/gatkdocs/org_broadinstitute_sting_gatk_walkers_annotator_QualByDepth.html">GATK's QualByDepth annotation</a> and default filter of QD &lt; 2.0, we incorporated a generalized version of this into <a href="https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/vfilter.py#L75">bcbio-nextgen's FreeBayes filter</a>: QUAL &lt; (depth-cutoff * 2.0) and DP &gt; depth-cutoff. </p>
</p></div>
</p></div>
<div id="outline-container-sec-4" class="outline-2">
<h2 id="sec-4">GATK variant quality score recalibration (VQSR)</h2>
<div class="outline-text-2" id="text-4">
<p> The other area where we needed to improve was using GATK Variant Quality Score Recalibration. The default parameters provide a set of calls that are overly conservative relative to the FreeBayes calls. VQSR provides the ability to tune the filtering so we experimented with multiple configurations to achieve approximately equal sensitivity relative to FreeBayes for both SNPs and Indels. The comparisons use the Genome in a Bottle reference callset for evaluation, and include VQSR default settings, multiple tranche levels and GATK's suggested hard filters: </p>
<p>  <a href="http://i.imgur.com/c4yWTMv.png">   <img src="http://i.imgur.com/c4yWTMv.png" width="700" alt="VQSR tuning: SNPs" /> </a>  <a href="http://i.imgur.com/3yZnkA6.png">   <img src="http://i.imgur.com/3yZnkA6.png" width="700" alt="VQSR tuning: indels" /> </a>
<p> While the sensitivity/specificity tradeoff depends on the research question, in trying to set a generally useful default we'd like to be less conservative than the GATK VQSR default. We learned these tips and tricks for tuning VQSR filtering: </p>
<ul class="org-ul">
<li>The default setting for VQSR is not a tranche level (like 99.0), but rather a LOD score of 0. In this experiment, that corresponded to a tranche of ~99.0 for SNPs and ~98.0 for indels. The best-practice example documentation uses command line parameter that specify a consistent tranche of 99.0 for both SNPs and indels, so depending on which you follow as a default you'll get different sensitivities. </li>
<li>To increase sensitivity, increase the tranche level. My expectations were that decreasing the tranche level would include more variants, but that actually applies additional filters. My suggestion for understanding tranche levels is that they specify the percentage of variants you want to capture; a tranche of 99.9% captures 99.9% of the true cases in the training set, while 99.0% captures less. </li>
<li>We found tranche settings of 99.97% for SNPs and 98.0% for indels correspond to roughly the sensitivity/specificity that you achieve with FreeBayes. These are the new default settings in bcbio-nextgen. </li>
<li>Using <a href="https://github.com/chapmanb/bcbio-nextgen/blob/master/bcbio/variation/vfilter.py#L125">hard filtering of variants based on GATK recommendations</a> performs well and is also a good default choice. For SNPs, the hard filter defaults are less conservative and more in line with FreeBayes results than VQSR defaults. VQSR has improved specificity at the same sensitivity and has the advantage of being configurable, but will require an extra tuning step. </li>
</ul>
<p> Overall VQSR provides good filtering and the ability to tune sensitivity but requires validation work to select tranche cutoffs that are as sensitive as hard filter defaults, since default values tend to be overly conservative for SNP calling. In the absence of the ability or desire to tune VQSR tranche levels, the GATK hard filters provide a nice default without much of a loss in precision. </p>
</p></div>
</p></div>
<div id="outline-container-sec-5" class="outline-2">
<h2 id="sec-5">Data availability and future work</h2>
<div class="outline-text-2" id="text-5">
<p> Thanks to continued community work on improving variant calling evaluations, this post demonstrates practical improvements in bcbio-nextgen variant calling. We welcome interested contributors to re-run and expand on the analysis, with full instructions in the <a href="https://bcbio-nextgen.readthedocs.org/en/latest/contents/testing.html#whole-genome-trio-50x">bcbio-nextgen example pipeline documentation</a>. Some of the output files from the analysis may also be useful: </p>
<ul class="org-ul">
<li>VCF files for FreeBayes <a href="https://s3.amazonaws.com/bcbio_nextgen/comparison3/freebayes-tp-het.vcf.gz">true positive</a> and <a href="https://s3.amazonaws.com/bcbio_nextgen/comparison3/freebayes-fp-het.vcf.gz">false positive</a> heterozygote calls, used here to improve filtering via assessment of high depth regions. Heterozygotes make up the majority of false positive calls so take the most work to correctly filter and detect. </li>
<li><a href="https://s3.amazonaws.com/bcbio_nextgen/comparison3/giab-comparision-shared-gatk-fb-extras.vcf.gz">Shared false positives from FreeBayes and GATK HaplotypeCaller</a>. These are potential missing variants in the Genome in a Bottle reference. Alternatively, they may represent persistent errors found in multiple callers. </li>
</ul>
<p> We plan to continue to explore variant calling improvements in bcbio-nextgen. Our next steps are to use the trio population framework to compared pooled population calling versus <a href="http://gatkforums.broadinstitute.org/discussion/3896/the-gatk-reference-model-pipeline-for-incremental-joint-discovery-in-full-detail">the incremental joint discovery approach introduced in GATK 3</a>. We'd also like to compare with single sample calling followed by <a href="https://github.com/chapmanb/bcbio.variation.recall">subsequent squaring off/backfilling</a> to assess the value of concurrent population calling. We welcome suggestions and thoughts on this work and future directions. </p>
</p></div>
</p></div>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2014/03/06/improving-reproducibility-and-installation-of-genomic-analysis-pipelines-with-docker/" class="btn" title="Improving reproducibility and installation of genomic analysis pipelines with Docker">Previous</a>
      
      
        <a href="http://localhost:4000/2014/08/12/validated-whole-genome-structural-variation-detection-using-multiple-callers/" class="btn" title="Validated whole genome structural variation detection using multiple callers">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
