<!doctype html>
<!--[if lt IE 7]><html class="no-js lt-ie9 lt-ie8 lt-ie7" lang="en"> <![endif]-->
<!--[if (IE 7)&!(IEMobile)]><html class="no-js lt-ie9 lt-ie8" lang="en"><![endif]-->
<!--[if (IE 8)&!(IEMobile)]><html class="no-js lt-ie9" lang="en"><![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en"><!--<![endif]-->
<head>
<meta charset="utf-8">
<title>Differential expression analysis with Bioconductor and Python &#8211; Blue Collar Bioinformatics</title>

<meta name="keywords" content="how-to">

  

<!-- Twitter Cards -->
<meta name="twitter:title" content="Differential expression analysis with Bioconductor and Python">

<meta name="twitter:site" content="@chapmanb">
<meta name="twitter:creator" content="@chapmanb">

<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="http://localhost:4000/images/">

<!-- Open Graph -->
<meta property="og:locale" content="en_US">
<meta property="og:type" content="article">
<meta property="og:title" content="Differential expression analysis with Bioconductor and Python">

<meta property="og:url" content="http://localhost:4000/2009/09/13/differential-expression-analysis-with-bioconductor-and-python/">
<meta property="og:site_name" content="Blue Collar Bioinformatics">





<link rel="canonical" href="http://localhost:4000/2009/09/13/differential-expression-analysis-with-bioconductor-and-python/">
<link href="http://feeds2.feedburner.com/bcbio" type="application/atom+xml" rel="alternate" title="Blue Collar Bioinformatics Feed">

<!-- https://t.co/dKP3o1e -->
<meta name="HandheldFriendly" content="True">
<meta name="MobileOptimized" content="320">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<!-- For all browsers -->
<link rel="stylesheet" href="http://localhost:4000/assets/css/main.css">
<!-- Webfonts -->
<script src="https://use.edgefonts.net/source-sans-pro:n2,i2,n3,i3,n4,i4,n6,i6,n7,i7,n9,i9;source-code-pro:n4,n7;volkhov.js"></script>

<meta http-equiv="cleartype" content="on">

<!-- HTML5 Shiv and Media Query Support -->
<!--[if lt IE 9]>
  <script src="http://localhost:4000/assets/js/vendor/html5shiv.min.js"></script>
  <script src="http://localhost:4000/assets/js/vendor/respond.min.js"></script>
<![endif]-->

<!-- Modernizr -->
<script src="http://localhost:4000/assets/js/vendor/modernizr-2.7.1.custom.min.js"></script>


<!-- MathJax -->
<script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>


<!-- Icons -->
<!-- 16x16 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.ico">
<!-- 32x32 -->
<link rel="shortcut icon" href="http://localhost:4000/favicon.png">
<!-- 57x57 (precomposed) for iPhone 3GS, pre-2011 iPod Touch and older Android devices -->
<link rel="apple-touch-icon-precomposed" href="http://localhost:4000/images/apple-touch-icon-precomposed.png">
<!-- 72x72 (precomposed) for 1st generation iPad, iPad 2 and iPad mini -->
<link rel="apple-touch-icon-precomposed" sizes="72x72" href="http://localhost:4000/images/apple-touch-icon-72x72-precomposed.png">
<!-- 114x114 (precomposed) for iPhone 4, 4S, 5 and post-2011 iPod Touch -->
<link rel="apple-touch-icon-precomposed" sizes="114x114" href="http://localhost:4000/images/apple-touch-icon-114x114-precomposed.png">
<!-- 144x144 (precomposed) for iPad 3rd and 4th generation -->
<link rel="apple-touch-icon-precomposed" sizes="144x144" href="http://localhost:4000/images/apple-touch-icon-144x144-precomposed.png">

</head>

<body id="post">

<div class="navigation-wrapper">
	<nav role="navigation" id="site-nav" class="animated drop">
	    <ul>
      
		    
		        
		    
		    <li><a href="http://localhost:4000/" >About</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/articles/" >Recent</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/tags/" >Topics</a></li>
		  
		    
		        
		    
		    <li><a href="http://localhost:4000/search/" >Search</a></li>
		  
	    </ul>
	</nav>
</div><!-- /.navigation-wrapper -->

<!--[if lt IE 9]><div class="upgrade"><strong><a href="http://whatbrowser.org/">Your browser is quite old!</strong> Why not upgrade to a different browser to better enjoy this site?</a></div><![endif]-->

<header class="masthead">
	<div class="wrap">
      
      <h1 class="site-title animated fadeIn"><a href="http://localhost:4000/">Blue Collar Bioinformatics</a></h1>
		<h2 class="site-description animated fadeIn" itemprop="description">Biological data analysis, validation and visualization</h2>
	</div>
</header><!-- /.masthead -->

<div class="js-menu-screen menu-screen"></div>


<div id="main" role="main">
  <article class="hentry">
    
    <div class="entry-wrapper">
      <header class="entry-header">
        <span class="entry-tags"><a href="http://localhost:4000/tags/#how-to" title="Pages tagged how-to">how-to</a></span>
        
          <h1 class="entry-title">Differential expression analysis with Bioconductor and Python</h1>
        
      </header>
      <footer class="entry-meta">
        
          
        
          <img src="http://localhost:4000/images/diego-work.jpg" class="bio-photo" alt="Brad Chapman bio photo"></a>
        
        <span class="author vcard">By <span class="fn">Brad Chapman</span></span>
        <span class="entry-date date published"><time datetime="2009-09-13T20:28:19-04:00"><i class="fa fa-calendar-o"></i> September 13, 2009</time></span>
        
        <span class="entry-comments"><i class="fa fa-comment-o"></i> <a href="#disqus_thread">Comment</a></span>
        
        
      </footer>
      <div class="entry-content">
        <p>
This post demonstrates performing differential expression analysis of short read sequencing data using a combination of Python and the <a href="http://www.r-project.org/">R statistical language</a>. Python is used a glue language to manipulate and prepare count data from short read sequencing. R and the <a href="http://www.bioconductor.org/">Bioconductor</a> package are used to perform the statistical analysis. The excellent <a href="http://rpy.sourceforge.net/rpy2.html">rpy2</a> package connection Python and R.</p>
<p>
Scientists often need to compare expression results from multiple experimental conditions. This can be done with <a href="http://en.wikipedia.org/wiki/DNA_microarray">microarray analysis</a> and the wide variety of associated <a href="http://www.bioconductor.org/packages/devel/Microarray.htm">statistical tests</a>, but many researchers are now utilizing short read sequencing. Experiments identify transcript prevalence through <a href="http://www.illumina.com/pages.ilmn?ID=278">digital gene expression</a> under multiple conditions, and we are interested in extracting statistically meaningful differences from the results. Similarly, we may be interested in differences resulting from <a href="http://www.illumina.com/pages.ilmn?ID=277">short RNA discovery</a> or other next generation sequencing applications.</p>
<p>
Short read differential expression analysis differs from microarray analyses as it is based on counts instead of intensity scores. The <a href="http://www.bioconductor.org/packages/devel/bioc/html/edgeR.html">edgeR</a> package in Bioconductor fits count data to a statistical model considering sample to sample variation, and performs <a href="http://en.wikipedia.org/wiki/Fisher's_exact_test">Fisher's exact test</a> to provide p-values associated with changes in expression between samples.</p>
<p>
Here we will consider the case of a two experiment analysis. A simple <a href="http://github.com/chapmanb/bcbb/blob/master/stats/diffexp_example.csv">Example CSV file</a> has counts for 5 different genes under 2 conditions and total count information for the full experiment. We read the file into a dictionary of counts keyed by each condition and the gene name:</p>
<p>[sourcecode language="python"]<br />
import csv<br />
import collections</p>
<p>def read_count_file(in_file):<br />
    """Read count information from a simple CSV file into a dictionary.<br />
    """<br />
    counts = collections.defaultdict(dict)<br />
    with open(in_file) as in_handle:<br />
        reader = csv.reader(in_handle)<br />
        header = reader.next()<br />
        conditions = header[1:]<br />
        for parts in reader:<br />
            region_name = parts[0]<br />
            region_counts = [float(x) for x in parts[1:]]<br />
            for ci, condition in enumerate(conditions):<br />
                counts[condition][region_name] = region_counts[ci]<br />
    return dict(counts)<br />
[/sourcecode]</p>
<p>
The dictionary is organized into <a href="http://numpy.scipy.org/">NumPy</a> matrices; NumPy is a powerful numerical package for Python which integrates smoothly with our rpy2 interface to import directly into R. Here we organize our conditions and genes and then push the count data into a matrix where the columns are conditions and the rows are genes; each item is a count value. This is returned along with associated data:</p>
<ul>
<li>groups -- A list of experimental groups, which can be used to analyze replicates. Here, two groups with a single replicate each are defined.
  </li>
<li>sizes -- The total number of counts for each experiment. This is extracted from the "Total" row of the CSV count file, and will be used for normalization of during the statistical analysis.
</li>
</ul>
<p>[sourcecode language="python"]<br />
import numpy</p>
<p>def get_conditions_and_genes(work_counts):<br />
    conditions = work_counts.keys()<br />
    conditions.sort()<br />
    all_genes = []<br />
    for c in conditions:<br />
        all_genes.extend(work_counts[c].keys())<br />
    all_genes = list(set(all_genes))<br />
    all_genes.sort()<br />
    sizes = [work_counts[c]["Total"] for c in conditions]<br />
    all_genes.remove("Total")<br />
    return conditions, all_genes, sizes</p>
<p>def edger_matrices(work_counts):<br />
    conditions, all_genes, sizes = get_conditions_and_genes(work_counts)<br />
    assert len(sizes) == 2<br />
    groups = [1, 2]<br />
    data = []<br />
    final_genes = []<br />
    for g in all_genes:<br />
        cur_row = [int(work_counts[c][g]) for c in conditions]<br />
        if sum(cur_row) > 0:<br />
            data.append(cur_row)<br />
            final_genes.append(g)<br />
    return (numpy.array(data), numpy.array(groups), numpy.array(sizes),<br />
            conditions, final_genes)<br />
[/sourcecode]</p>
<p>
The organized data is now ready to be pushed directly into an edgeR Bioconductor analysis using the rpy2 interface. Three R functions are called: the data matrices are organized into a <code>DGEList</code> object, this object is passed to <code>deDGE</code> which does the actual digital gene expression analysis, and finally <code>topTags</code> is called to retrieve a vector of differential expression p-values. The vector is translated into a Python object from which we extract the p-values and re-organize them by the initial gene indexes. The ordered p-values are then returned.</p>
<p>[sourcecode language="python"]<br />
import rpy2.robjects as robjects<br />
import rpy2.robjects.numpy2ri</p>
<p>def run_edger(data, groups, sizes, genes):<br />
    robjects.r('''<br />
        library(edgeR)<br />
    ''')<br />
    params = {'group' : groups, 'lib.size' : sizes}<br />
    dgelist = robjects.r.DGEList(data, **params)<br />
    ms = robjects.r.deDGE(dgelist, doPoisson=True)<br />
    tags = robjects.r.topTags(ms, pair=groups, n=len(genes))<br />
    indexes = [int(t) - 1 for t in tags.rownames()]<br />
    pvals = list(tags.r['adj.P.Val'][0])<br />
    assert len(indexes) == len(pvals)<br />
    pvals_w_index = zip(indexes, pvals)<br />
    pvals_w_index.sort()<br />
    assert len(pvals_w_index) == len(indexes)<br />
    return [p for i,p in pvals_w_index]<br />
[/sourcecode]</p>
<p>
The final results are written to a CSV file with our ordered genes, conditions and p-value probabilities. Each gene is associated with the initial count data and p-values, the genes are sorted by p-value with the most differentially expressed genes placed first, and the ordered information is written out:</p>
<p>[sourcecode language="python"]<br />
def write_outfile(outfile, genes, conditions, work_counts, probs):<br />
    with open(outfile, "w") as out_handle:<br />
        writer = csv.writer(out_handle)<br />
        writer.writerow(["Region"] +<br />
                ["%s count" % c for c in conditions] + ["edgeR p-value"])<br />
        out_info = []<br />
        for i, gene in enumerate(genes):<br />
            counts = [int(work_counts[c][gene]) for c in conditions]<br />
            out_info.append((probs[i], [gene] + counts))<br />
        out_info.sort()<br />
        [writer.writerow(start + [prob]) for prob, start in out_info]<br />
[/sourcecode]</p>
<p>
Our <a href="http://github.com/chapmanb/bcbb/blob/master/stats/diffexp_example-diffs.csv">example output file</a> shows the results. The <a href="http://github.com/chapmanb/bcbb/blob/master/stats/count_diffexp.py">full script</a> is available for you to use and customize for your own analyses. This example demonstrates the power of the rpy2 interface. Input and output data is happily manipulated in a familiar language such as Python, and can be seamlessly integrated with excellent statistical computation in Bioconductor and other specialized R packages.</p>

        
          <div id="disqus_thread"></div><!-- /#disqus_thread -->
          
<script type="text/javascript">
    /* * * CONFIGURATION VARIABLES: EDIT BEFORE PASTING INTO YOUR WEBPAGE * * */
    var disqus_shortname = 'bcbio'; // required: replace example with your forum shortname

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function() {
        var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
        dsq.src = 'https://' + disqus_shortname + '.disqus.com/embed.js';
        (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();

    /* * * DON'T EDIT BELOW THIS LINE * * */
    (function () {
        var s = document.createElement('script'); s.async = true;
        s.type = 'text/javascript';
        s.src = 'https://' + disqus_shortname + '.disqus.com/count.js';
        (document.getElementsByTagName('HEAD')[0] || document.getElementsByTagName('BODY')[0]).appendChild(s);
    }());
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>

        
      </div><!-- /.entry-content -->
    </div><!-- /.entry-wrapper -->
    <nav class="pagination" role="navigation">
      
        <a href="http://localhost:4000/2009/09/07/usage-plans-for-amazon-web-services-research-grant/" class="btn" title="Usage plans for Amazon Web Services research grant">Previous</a>
      
      
        <a href="http://localhost:4000/2009/10/18/gene-ontology-analysis-with-python-and-bioconductor/" class="btn" title="Gene Ontology analysis with Python and Bioconductor">Next</a>
      
    </nav><!-- /.pagination -->
  </article>
</div><!-- /#main -->

<div class="footer-wrapper">
  <footer role="contentinfo" class="entry-wrapper">
    

<div class="social-icons">
	<a href="https://twitter.com/chapmanb" title="Brad Chapman on Twitter" target="_blank"><i class="fa fa-twitter-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Github" target="_blank"><i class="fa fa-github-square fa-2x"></i></a>
	<a href="https://github.com/chapmanb" title="Brad Chapman on Bitbucket" target="_blank"><i class="fa fa-bitbucket-square fa-2x"></i></a>
	
	
	
	
	
	
  
	
	<a href="https://linkedin.com/pub/3/300/8ba" title="Brad Chapman on LinkedIn" target="_blank"><i class="fa fa-linkedin-square fa-2x"></i></a>
  <a href="http://feeds2.feedburner.com/bcbio" title="Atom/RSS feed"><i class="fa fa-rss-square fa-2x"></i></a>
</div><!-- /.social-icons -->

  </footer>
</div><!-- /.footer-wrapper -->

<script type="text/javascript">
  var BASE_URL = 'http://localhost:4000';
</script>

<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script>window.jQuery || document.write('<script src="http://localhost:4000/assets/js/vendor/jquery-1.9.1.min.js"><\/script>')</script>
<script src="http://localhost:4000/assets/js/scripts.min.js"></script>




</body>
</html>
